{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# getting SNP and coverage data from bam files or pileup files\n",
    "## python wrapper around samtools / mawktools\n",
    "+ bam --> mpileup --> clean --> filter --> pile2SNP\n",
    "+ mpileup --> clean --> filter --> pile2SNP\n",
    "+ cleanpileup --> filter --> pile2SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-19T07:43:59.830066Z",
     "start_time": "2021-05-19T07:43:59.824811Z"
    }
   },
   "outputs": [],
   "source": [
    "# get the code\n",
    "import sys\n",
    "import os\n",
    "sys.path.append('../code')\n",
    "\n",
    "\n",
    "# HOME\n",
    "home = '/Users/mahtin'\n",
    "home = '/Users/martinscience'\n",
    "\n",
    "# standard paths\n",
    "static = os.path.join(home, \"Dropbox/Icke/Work/static\")\n",
    "cluster_path = os.path.join(home, \"mount\")\n",
    "tooldata = os.path.join(home, \"Dropbox/Icke/Work/somVar/tooldata\")\n",
    "testdata = os.path.join(home,\"Dropbox/Icke/Work/somVar/testdata\")\n",
    "\n",
    " \n",
    "cnvdata = os.path.join(tooldata, \"myCNVdata\")\n",
    "output_path = os.path.join(cnvdata, \"output\")\n",
    "\n",
    "# the path to the input data\n",
    "cnv_path = os.path.join(cnvdata, \"cnv\")\n",
    "PONCOV_path = cnv_path = os.path.join(cnvdata, \"chromCov\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### load config and basic functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-19T10:57:06.796525Z",
     "start_time": "2021-05-19T10:57:06.787798Z"
    }
   },
   "outputs": [],
   "source": [
    "from script_utils_CNV import show_output, cmd2df, get_CNVconfig\n",
    "\n",
    "def get_rawCNV(\n",
    "    normal_bam=\"\",\n",
    "    tumor_bam=\"\",\n",
    "    tumor_bams=[],\n",
    "    TN_pileup_file=\"\",    # mpileup from normalbam, tumorbam(s)\n",
    "    clean_TN_pileup_file=\"\", # mpileup from normalbam, tumorbam(s) with cleanpileup.mawk already done\n",
    "    pileup_is_clean=True,\n",
    "    chrom=\"\",\n",
    "    config=CNVconfig,\n",
    "    SNP_output=\"\"\n",
    "    ):\n",
    "    \n",
    "    ## PARAMS\n",
    "    # unwrap mawk tools\n",
    "    def mawk(tool):\n",
    "        return os.path.join(config[\"mawk_path\"], f\"{tool}.mawk\")\n",
    "\n",
    "    sc = config['hetSNP']\n",
    "    cc = config['coverage']\n",
    "    \n",
    "    \n",
    "    # create the basic command and unpack required params\n",
    "    ###### FILTERBED\n",
    "    filter_cmd = f\"{mawk('filterBed')} {config['bed_file']} -x -c {chrom}\"\n",
    "    \n",
    "    if not SNP_output:\n",
    "        show_output(\"Output file for heteroSNP is missing!\", color=\"warning\")\n",
    "        return\n",
    "    ###### PILE2CNV \n",
    "    cnv_cmd = f\"{mawk('pile2CNV')} -x -o {SNP_output} -v {sc['normalVAF'][0]} -V {sc['normalVAF'][1]} -d {sc['minDepth']} -c {cc['minCov']}\"\n",
    "    \n",
    "    # combine \n",
    "    cmd = f\"{filter_cmd} | {cnv_cmd}\"\n",
    "    \n",
    "    \n",
    "    if clean_TN_pileup_file:\n",
    "        if os.path.splitext(clean_TN_pileup_file)[1] == \".gz\":\n",
    "            cmd = f\"gunzip < {clean_TN_pileup_file} | {cmd}\"\n",
    "        else:\n",
    "            cmd = f\"cat {clean_TN_pileup_file} | {cmd}\"\n",
    "    else:\n",
    "        # add the cleanup cmd if \n",
    "        cmd = f\"{mawk('cleanpileup')} -d | {cmd}\"\n",
    "        if TN_pileup_file:\n",
    "            if os.path.splitext(TN_pileup_file)[1] == \".gz\":\n",
    "                cmd = f\"gunzip < {TN_pileup_file} | {cmd}\"\n",
    "            else:\n",
    "                cmd = f\"cat {TN_pileup_file} | {cmd}\"\n",
    "        else:\n",
    "            # pileup has to be done\n",
    "            # get the bam files\n",
    "            if tumor_bams:\n",
    "                bams = \" \".join([normal_bam], tumor_bams)\n",
    "            else:\n",
    "                bams = f\"{normal_bam} {tumor_bam}\"\n",
    "            # get params from config\n",
    "            pc = config['pileup']\n",
    "            split_genome = os.path.join(config['genome_split_path'], f\"{chrom}.fa\")\n",
    "            pileup_cmd = f\"samtools mpileup -f {split_genome} -l {config['bed_file']} -r {chrom} -q {pc['MAPQ']} -Q {pc['Q']} {bams}\"\n",
    "            cmd = f\"{pileup_cmd} | {cmd}\"\n",
    "    try:\n",
    "        cov_df = cmd2df(cmd, show=True, multi=False)\n",
    "        return cov_df\n",
    "    except:\n",
    "        show_output(\"There was an error using shell command\", color=\"warning\")\n",
    "        \n",
    "    \n",
    "\n",
    "\n",
    "    # matrix_df = cmd2df(cmd)\n",
    "    # return matrix_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A) get raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-19T10:27:26.536577Z",
     "start_time": "2021-05-19T10:27:03.755171Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m$ gunzip < /Users/martinscience/Dropbox/Icke/Work/somVar/tooldata/myCNVdata/cleanpileup/02_A-B.chr1.gz | ../shell/filterBed.mawk /Users/martinscience/Dropbox/Icke/Work/static/bed_files/SureSelect/hg38/SS_HAEv7_hg38_Padded.bed -x -c chr1 | ../shell/pile2CNV.mawk -x -o /Users/martinscience/Dropbox/Icke/Work/somVar/tooldata/myCNVdata/output/pile2CNV/02_A-B.chr7.snp -v 0.2 -V 0.8 -d 15 -c 30\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>ExonPos</th>\n",
       "      <th>Cov1</th>\n",
       "      <th>Cov2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16760</td>\n",
       "      <td>4630</td>\n",
       "      <td>26.08</td>\n",
       "      <td>35.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16770</td>\n",
       "      <td>4640</td>\n",
       "      <td>30.39</td>\n",
       "      <td>42.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16780</td>\n",
       "      <td>4650</td>\n",
       "      <td>34.31</td>\n",
       "      <td>48.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16790</td>\n",
       "      <td>4660</td>\n",
       "      <td>37.49</td>\n",
       "      <td>54.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16800</td>\n",
       "      <td>4670</td>\n",
       "      <td>39.99</td>\n",
       "      <td>58.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578119</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918350</td>\n",
       "      <td>248906220</td>\n",
       "      <td>24037.60</td>\n",
       "      <td>28822.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578120</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918360</td>\n",
       "      <td>248906230</td>\n",
       "      <td>24035.90</td>\n",
       "      <td>28821.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578121</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918370</td>\n",
       "      <td>248906240</td>\n",
       "      <td>24034.30</td>\n",
       "      <td>28819.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578122</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918380</td>\n",
       "      <td>248906250</td>\n",
       "      <td>24032.80</td>\n",
       "      <td>28817.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578123</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918390</td>\n",
       "      <td>248906260</td>\n",
       "      <td>24031.20</td>\n",
       "      <td>28815.60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>578124 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chr      Start    ExonPos      Cov1      Cov2\n",
       "0       chr1      16760       4630     26.08     35.48\n",
       "1       chr1      16770       4640     30.39     42.44\n",
       "2       chr1      16780       4650     34.31     48.84\n",
       "3       chr1      16790       4660     37.49     54.29\n",
       "4       chr1      16800       4670     39.99     58.71\n",
       "...      ...        ...        ...       ...       ...\n",
       "578119  chr1  248918350  248906220  24037.60  28822.50\n",
       "578120  chr1  248918360  248906230  24035.90  28821.00\n",
       "578121  chr1  248918370  248906240  24034.30  28819.40\n",
       "578122  chr1  248918380  248906250  24032.80  28817.60\n",
       "578123  chr1  248918390  248906260  24031.20  28815.60\n",
       "\n",
       "[578124 rows x 5 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " CNVconfig = get_CNVconfig(\n",
    "    \"../config/config_CNV.yaml\", \n",
    "    local_config=dict(\n",
    "        mawk_path=\"../shell\",\n",
    "        bed_file=os.path.join(static, \"bed_files/SureSelect/hg38/SS_HAEv7_hg38_Padded.bed\"),\n",
    "        genome_split_path=os.path.join(static, \"genome/gatk/hg38/split\")\n",
    "    ))\n",
    "\n",
    "get_rawCNV(\n",
    "    clean_TN_pileup_file=os.path.join(cnvdata, \"cleanpileup/02_A-B.chr1.gz\"),\n",
    "    chrom=\"chr1\",\n",
    "    SNP_output=os.path.join(output_path, \"pile2CNV/03_A-B.chr7.snp\"),\n",
    "    config=CNVconfig\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-19T10:30:23.774260Z",
     "start_time": "2021-05-19T10:28:07.001533Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m$ gunzip < /Users/martinscience/Dropbox/Icke/Work/somVar/tooldata/myCNVdata/cleanpileup/03_A-B.chr1.gz | ../shell/filterBed.mawk /Users/martinscience/Dropbox/Icke/Work/static/bed_files/SureSelect/hg38/SS_HAEv7_hg38_Padded.bed -x -c chr1 | ../shell/pile2CNV.mawk -x -o /Users/martinscience/Dropbox/Icke/Work/somVar/tooldata/myCNVdata/output/pile2CNV/03_A-B.chr1.snp -v 0.2 -V 0.8 -d 15 -c 30\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>ExonPos</th>\n",
       "      <th>Cov1</th>\n",
       "      <th>Cov2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13410</td>\n",
       "      <td>1647</td>\n",
       "      <td>20.13</td>\n",
       "      <td>39.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13420</td>\n",
       "      <td>1657</td>\n",
       "      <td>27.65</td>\n",
       "      <td>54.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13430</td>\n",
       "      <td>1667</td>\n",
       "      <td>36.32</td>\n",
       "      <td>70.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13440</td>\n",
       "      <td>1677</td>\n",
       "      <td>45.66</td>\n",
       "      <td>88.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13450</td>\n",
       "      <td>1687</td>\n",
       "      <td>55.68</td>\n",
       "      <td>106.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581995</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918350</td>\n",
       "      <td>248906587</td>\n",
       "      <td>239876.00</td>\n",
       "      <td>287575.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581996</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918360</td>\n",
       "      <td>248906597</td>\n",
       "      <td>239862.00</td>\n",
       "      <td>287557.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581997</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918370</td>\n",
       "      <td>248906607</td>\n",
       "      <td>239847.00</td>\n",
       "      <td>287537.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581998</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918380</td>\n",
       "      <td>248906617</td>\n",
       "      <td>239830.00</td>\n",
       "      <td>287514.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581999</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248918390</td>\n",
       "      <td>248906627</td>\n",
       "      <td>239813.00</td>\n",
       "      <td>287489.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>582000 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chr      Start    ExonPos       Cov1       Cov2\n",
       "0       chr1      13410       1647      20.13      39.14\n",
       "1       chr1      13420       1657      27.65      54.46\n",
       "2       chr1      13430       1667      36.32      70.81\n",
       "3       chr1      13440       1677      45.66      88.10\n",
       "4       chr1      13450       1687      55.68     106.04\n",
       "...      ...        ...        ...        ...        ...\n",
       "581995  chr1  248918350  248906587  239876.00  287575.00\n",
       "581996  chr1  248918360  248906597  239862.00  287557.00\n",
       "581997  chr1  248918370  248906607  239847.00  287537.00\n",
       "581998  chr1  248918380  248906617  239830.00  287514.00\n",
       "581999  chr1  248918390  248906627  239813.00  287489.00\n",
       "\n",
       "[582000 rows x 5 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_rawCNV(\n",
    "    clean_TN_pileup_file=os.path.join(cnvdata, \"cleanpileup/03_A-B.chr1.gz\"),\n",
    "    chrom=\"chr1\",\n",
    "    SNP_output=os.path.join(output_path, \"pile2CNV/03_A-B.chr1.snp\"),\n",
    "    config=CNVconfig\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T10:57:12.978Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m$ samtools mpileup -f /Users/martinscience/Dropbox/Icke/Work/static/genome/gatk/hg38/split/chr7.fa -l /Users/martinscience/Dropbox/Icke/Work/static/bed_files/SureSelect/hg38/SS_HAEv7_hg38_Padded.bed -r chr7 -q 20 -Q 25 /Users/martinscience/Dropbox/Icke/Work/somVar/testdata/bam/03_B.chr7.bam /Users/martinscience/Dropbox/Icke/Work/somVar/testdata/bam/03_A.chr7.bam | ../shell/cleanpileup.mawk -d | ../shell/filterBed.mawk /Users/martinscience/Dropbox/Icke/Work/static/bed_files/SureSelect/hg38/SS_HAEv7_hg38_Padded.bed -x -c chr7 | ../shell/pile2CNV.mawk -x -o /Users/martinscience/Dropbox/Icke/Work/somVar/tooldata/myCNVdata/output/pile2CNV/03_A-B.chr7.snp -v 0.2 -V 0.8 -d 15 -c 30\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "get_rawCNV(\n",
    "    normal_bam=os.path.join(testdata, \"bam/03_B.chr7.bam\"),\n",
    "    tumor_bam=os.path.join(testdata, \"bam/03_A.chr7.bam\"),\n",
    "    chrom=\"chr7\",\n",
    "    SNP_output=os.path.join(output_path, \"pile2CNV/03_A-B.chr7.snp\"),\n",
    "    config=CNVconfig\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## B) coverage: merge sample coverage with Pon coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_Covdata(sample, sample_cnv_path=\"\", PON_cnv_path=\"\", verbose=False, filtered=True):\n",
    "    \n",
    "    cover_dfs = []\n",
    "    chrom_list = [f\"chr{chrom + 1}\" for chrom in range(22)] + ['chrX']\n",
    "    for chrom in chrom_list:\n",
    "        # reading sampleCoverage\n",
    "        sample_cov_file = os.path.join(sample_cnv_path, f\"{sample}.{chrom}.cov\")\n",
    "        if not os.path.isfile(sample_cov_file):\n",
    "            print(\"No file\", sample_cov_file)\n",
    "            continue\n",
    "        if verbose:\n",
    "            print(f\"Reading coverage from {chrom} of sample {sample} from {sample_cov_file}.\")\n",
    "        cov_df = pd.read_csv(sample_cov_file, sep='\\t', compression=\"gzip\")\n",
    "        \n",
    "        #reading PONcoverage\n",
    "        full_or_filtered = \"filtered\" if filtered else \"full\"\n",
    "        pon_cov_file = os.path.join(PON_cnv_path, f\"{chrom}.{full_or_filtered}.csv.gz\")\n",
    "        if not os.path.isfile(pon_cov_file):\n",
    "            print(\"No file\", pon_cov_file)\n",
    "            continue\n",
    "        if verbose:\n",
    "            print(f\"Reading PON coverage of {chrom} from file {pon_cov_file}.\")\n",
    "        pon_df = pd.read_csv(pon_cov_file, sep='\\t', compression=\"gzip\").loc[:,['Chr', 'Pos', 'FullExonPos', 'ExonPos', 'meanCov', 'medianCov', 'std']]\n",
    "        # column rename\n",
    "        trans_dict = {col:f\"PON{col}\" for col in pon_df.columns[4:]}\n",
    "        pon_df = pon_df.rename(columns=trans_dict)\n",
    "        # merge sample with PON coverage\n",
    "        sample_df = cov_df.merge(pon_df, on=['Chr', 'Pos', 'ExonPos'], how=\"outer\").loc[:,['Chr', 'Pos', 'FullExonPos', 'ExonPos', 'Coverage','PONmeanCov', 'PONmedianCov', 'PONstd']]\n",
    "        \n",
    "        ##### here recover missing FullExonPos from margin\n",
    "        # get \n",
    "        exon_start, full_start = sample_df.iloc[0][['ExonPos', 'FullExonPos']]\n",
    "        offset = full_start - exon_start\n",
    "        sample_df.loc[sample_df['FullExonPos'] != sample_df['FullExonPos'], 'FullExonPos'] = sample_df['ExonPos'] + offset\n",
    "        sample_df.loc[:, 'FullExonPos'] = sample_df.loc[:, 'FullExonPos'].astype(int)\n",
    "        cover_dfs.append(sample_df)  \n",
    "    # combine chrom data\n",
    "    cover_df = pd.concat(cover_dfs)\n",
    "    \n",
    "    # normalize the coverage\n",
    "    cover_df['Coverage'] = cover_df['Coverage'].fillna(0)\n",
    "    mean_cov = sample_df['Coverage'].mean()\n",
    "    cover_df.loc[:, 'Coverage'] = (cover_df['Coverage'] / mean_cov * 100)\n",
    "    # loggable are the coverages, where log2ratio can be computed\n",
    "    loggable = (cover_df['PONmeanCov'] * cover_df['Coverage'] != 0)\n",
    "    cover_df.loc[loggable, 'log2ratio'] = np.log2(cover_df.loc[loggable, 'Coverage'] / cover_df.loc[loggable, 'PONmeanCov'])\n",
    "    # mark regions without PON coverage as 0\n",
    "    cover_df.loc[~loggable, 'log2ratio'] = np.nan\n",
    "    return cover_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
