{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T16:13:29.643568Z",
     "start_time": "2020-10-30T16:13:29.601982Z"
    }
   },
   "outputs": [],
   "source": [
    "# HOME\n",
    "import os\n",
    "import scipy\n",
    "home = '/Users/mahtin'\n",
    "# home = '/Users/martinscience'\n",
    "# standard paths\n",
    "testdata = os.path.join(home,\"Dropbox/Icke/Work/somVar/testdata\")\n",
    "static_path = os.path.join(home, \"Dropbox/Icke/Work/static\")\n",
    "cluster_path = os.path.join(home, \"mount\")\n",
    "tooldata = os.path.join(home, \"Dropbox/Icke/Work/somVar/tooldata\")\n",
    "\n",
    "# tool-specific paths\n",
    "shell_path = \"../shell\"\n",
    "# \n",
    "cnvdata = os.path.join(tooldata, \"myCNVdata\")\n",
    "output_path = os.path.join(cnvdata, \"output\")\n",
    "\n",
    "# the path to the input data\n",
    "cnv_path = os.path.join(cnvdata, \"cnv\")\n",
    "cnvPON_path = cnv_path = os.path.join(cnvdata, \"chromCov\")\n",
    "cnv_path = os.path.join(cluster_path, \"scratch/develop/PONcoverage/cnv\")\n",
    "cnvPON_path = os.path.join(cluster_path, \"scratch/develop/PONcoverage/chromCov\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-28T09:08:46.582938Z",
     "start_time": "2020-10-28T09:08:46.579367Z"
    }
   },
   "outputs": [],
   "source": [
    "cnvPON_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## visualize coverage data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-28T09:08:48.838658Z",
     "start_time": "2020-10-28T09:08:48.075471Z"
    }
   },
   "outputs": [],
   "source": [
    "sample = \"01_A\"\n",
    "cov_df = pd.read_csv(os.path.join(output_path, f'rollingCNV/{sample}.rollcov.csv'), sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### visualize the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-28T09:08:49.732571Z",
     "start_time": "2020-10-28T09:08:49.729434Z"
    }
   },
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "# use seaborn plotting defaults\n",
    "import seaborn as sns; sns.set()\n",
    "from matplotlib.collections import PatchCollection\n",
    "from matplotlib.patches import Rectangle\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-28T10:10:50.102902Z",
     "start_time": "2020-10-28T10:10:49.970759Z"
    }
   },
   "outputs": [],
   "source": [
    "def sort_df(df):\n",
    "    '''\n",
    "    helper for sorting dfs for chromosomes\n",
    "    '''\n",
    "    df2 = df.copy()\n",
    "    # make Chr column categorical for sorting .. and sort\n",
    "    chrom_list = [f\"chr{i+1}\" for i in range(22)] + ['chrX']\n",
    "    df2['Chr'] = pd.Categorical(df2['Chr'], chrom_list)\n",
    "    return df2.sort_values(['Chr', 'FullExonPos'])\n",
    "\n",
    "\n",
    "def get_chrom_df(df):\n",
    "\n",
    "    # dropna is neccessary because grouping on categorical returns all categories\n",
    "    chrom_df = df.groupby('Chr')['FullExonPos'].agg(['mean', 'min', 'max']).dropna()\n",
    "    cols = list(chrom_df.columns)\n",
    "    chrom_df['sum'] = chrom_df['max'] - chrom_df['min']\n",
    "    chrom_df['cummin'] = chrom_df['sum'].cumsum()\n",
    "    chrom_df['dif'] = (chrom_df['max'] - chrom_df['cummin']).astype(int)\n",
    "    for col in cols:\n",
    "        chrom_df[col] = (chrom_df[col] - chrom_df['dif']).astype(int)\n",
    "    cols.append('dif')\n",
    "    return chrom_df.loc[:, cols]\n",
    "\n",
    "\n",
    "def make_color_chroms(ax, chrom_df, color_chroms, ylimits=(-10,10), colormap='coolwarm_r'):\n",
    "    \n",
    "    # set the cmap from provided argument\n",
    "    cmap = plt.cm.get_cmap(colormap, 23)\n",
    "    \n",
    "    # build the rects\n",
    "    rects = []\n",
    "    # set the height and ymin beyond the ylimits so borders are not seen\n",
    "    ymin = ylimits[0] * 1.1\n",
    "    height = (ylimits[1] - ymin) * 1.1\n",
    "    \n",
    "    for chrom in chrom_df.index:\n",
    "        row = chrom_df.loc[chrom]\n",
    "        rect = Rectangle((row['min'], ymin), width=row['max']-row['min'], height=height)\n",
    "        rects.append(rect)\n",
    "\n",
    "    if color_chroms:\n",
    "        rect_kwargs = dict(\n",
    "            alpha=.6,\n",
    "            ec='none'\n",
    "        )\n",
    "    else:\n",
    "        rect_kwargs = dict(\n",
    "            alpha=1,\n",
    "            fc='none',\n",
    "            ec='darkgray',\n",
    "            lw=1,\n",
    "            ls='-'\n",
    "        )\n",
    "    # set the rectangle collection with colormap\n",
    "    rect_collection = PatchCollection(rects, cmap=cmap, **rect_kwargs)\n",
    "    # set the index for the color map from chrom integers\n",
    "    rect_collection.set_array(chrom_df.index.str.replace('chr', '').str.replace('X', '23').astype(int))\n",
    "    return ax.add_collection(rect_collection)\n",
    "\n",
    "\n",
    "def add_chrom_labels(ax, chrom_df, ylimits=(-10,10)):\n",
    "    \n",
    "    # YOFFSET is the upper-relative y-position \n",
    "    YOFFSET = 0.9\n",
    "    # get the min_chrom_fraction from minimum chrom_size / whole stretch\n",
    "    min_chrom_frac = (chrom_df['max'] - chrom_df['min']).min() / chrom_df['max'].max()\n",
    "    chrom_size = min(50, max(15, 200*min_chrom_frac))\n",
    "    style = dict(size=chrom_size, color='#2f3832')\n",
    "    # set the height and ymin beyond the ylimits so borders are not seen\n",
    "    ypos = ylimits[0] + YOFFSET * (ylimits[1] - ylimits[0])\n",
    "    for chrom, row in chrom_df.iterrows():\n",
    "        if (len(chrom_df.index) > 12):\n",
    "            chrom = chrom.replace('chr', '')\n",
    "        ax.text(row['mean'], ypos, chrom, ha='center', **style)\n",
    "    \n",
    "\n",
    "def make_nice(position):\n",
    "    '''\n",
    "    takes position and returns closest multiple of 1, 2, 5 or 10\n",
    "    '''\n",
    "    # set nice values\n",
    "    nice_positions = np.array([1,2,2.5,5,10])\n",
    "    # get the 10s\n",
    "    power10 = np.power(10, np.floor(np.log10(position)))\n",
    "    # reduce to value between 1 and 10\n",
    "    num = position / power10\n",
    "    # find the closest nice position\n",
    "    base = nice_positions[np.argmin(np.abs(nice_positions / num - 1))]\n",
    "    return base * power10\n",
    "\n",
    "def get_tick_pos(tick_dist, chrom_df):\n",
    "    '''\n",
    "    return from chrom_df the evenly-spread (tick_dist) positions per chrom\n",
    "    '''\n",
    "    return [pos for _, row in chrom_df.iterrows() for pos in range(row['min'] + tick_dist, row['max'], tick_dist)]\n",
    "\n",
    "\n",
    "def str_pos(pos, df, precision=1):\n",
    "    '''\n",
    "    returns string representation of base position\n",
    "    on genomic coords\n",
    "    '''\n",
    "    pos = df.iloc[np.argmin(np.abs(df['PlotPos'] - pos))]['Pos']\n",
    "    # get the closest base power\n",
    "    power10 = int(np.round(np.log10(pos) / 3) * 3)\n",
    "    # get the base fraction\n",
    "    base = pos / np.power(10, power10)\n",
    "    if power10 == 9:\n",
    "        base = base * 1000\n",
    "        power10 = 6\n",
    "    if power10 == 6:\n",
    "        suff = 'Mb'\n",
    "    elif power10 == 3:\n",
    "        suff = 'kb'\n",
    "    base = round(base,precision)\n",
    "    if power10 == 0:\n",
    "        suff = 'b'\n",
    "        base = int(base)\n",
    "    return f\"{base}{suff}\"\n",
    "\n",
    "\n",
    "def set_ticks(ax, df, chrom_df, ticks=20, label_size=12):\n",
    "    '''\n",
    "    for a given tick number, set nicely spread ticks\n",
    "    '''\n",
    "    \n",
    "    ## determine optimale tick distance\n",
    "    # get the chrom_number\n",
    "    chrom_count = len(chrom_df.index)\n",
    "    # get the number of bases\n",
    "    stretch = chrom_df['max'][-1]\n",
    "    # set the number of desired ticks\n",
    "    major_tick_dist = int(stretch / (ticks + 1))\n",
    "    minor_tick_dist = int(stretch / ((ticks * 2) + 1))\n",
    "    \n",
    "    # feed tick distance into chrom_df to get chrom-specific coords\n",
    "    major_pos = get_tick_pos(major_tick_dist, chrom_df)\n",
    "    minor_pos = [pos - minor_tick_dist for pos in major_pos]\n",
    "    \n",
    "    ax.xaxis.set_major_locator(plt.FixedLocator(major_pos))\n",
    "    # only print the genomic coords below a certain base total\n",
    "    if stretch < 2e7:\n",
    "        major_labels = [str_pos(pos, df) for pos in major_pos]\n",
    "        ax.xaxis.set_major_formatter(plt.FixedFormatter(major_labels))\n",
    "    else:\n",
    "        ax.xaxis.set_major_formatter(plt.NullFormatter())\n",
    "    ax.xaxis.set_minor_locator(plt.FixedLocator(minor_pos))\n",
    "    ax.xaxis.grid(which='major', linestyle='-', linewidth=2)\n",
    "    ax.xaxis.grid(which='minor', linestyle='--', linewidth=1)\n",
    "    ax.xaxis.set_tick_params(which='major', length=20, labelsize=label_size)\n",
    "    ax.yaxis.set_tick_params(which='major', length=20, labelsize=label_size)\n",
    "    # set the tick labels\n",
    "    for tick in ax.xaxis.get_majorticklabels():\n",
    "        tick.set_verticalalignment(\"bottom\")\n",
    "    return ax\n",
    "\n",
    "\n",
    "def extract_pos(region):\n",
    "    \n",
    "    def convert(pos):\n",
    "        if pos.endswith('Mb'):\n",
    "            pos = int(pos.replace('Mb', '')) * 1e6\n",
    "        elif pos.endswith('kb'):\n",
    "            pos = int(pos.replace('kb', '')) * 1000\n",
    "        else:\n",
    "            pos = int(pos)\n",
    "        return pos\n",
    "    \n",
    "    split = region.split(':')  \n",
    "    chrom = split[0]\n",
    "    \n",
    "    # if start and are used\n",
    "    if len(split) > 1 and '-' in split[1]:     \n",
    "        se = split[1].split('-')\n",
    "        start = convert(se[0])\n",
    "        end = convert(se[1])\n",
    "    else:\n",
    "        start = 0\n",
    "        end = 1e10\n",
    "    return chrom, start, end\n",
    "    \n",
    "\n",
    "def plot_genomic(df, plots, chroms='all', color_chroms=True, colormap='coolwarm_r', region='', figsize=(20,4), ylim=(-1,1), label_size=12):\n",
    "    \n",
    "    #### DATA MANGELING ##########\n",
    "    # get cols for rearranging\n",
    "    org_cols = list(df.columns)\n",
    "    \n",
    "    # sort the df\n",
    "    df = sort_df(df)\n",
    "    # reduce the df to the selected chromosomes\n",
    "    if region:\n",
    "        chrom, start, end = extract_pos(region)\n",
    "        df = df.query('Chr == @chrom and @start <= Pos <= @end')\n",
    "    elif chroms != 'all':\n",
    "        df = df.query('Chr in @chroms')\n",
    "\n",
    "    # get the chrom_df for collapsing the \n",
    "    chrom_df = get_chrom_df(df)\n",
    "    \n",
    "    df = df.merge(chrom_df.loc[:, 'dif'], on='Chr')\n",
    "    df['PlotPos'] = df['FullExonPos'] - df['dif']\n",
    "    \n",
    "    # rearrange the df as return value\n",
    "    new_cols = org_cols[:4] + ['PlotPos'] + org_cols[4:]\n",
    "    df = df.loc[:, new_cols]\n",
    "    \n",
    "\n",
    "    ######## PLOTTING #######\n",
    "    # plot the figure\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "    # set the x-axis limits\n",
    "    _ = ax.set_xlim(0, df['PlotPos'].max())\n",
    "    \n",
    "    # plot the graphs #######\n",
    "    for plot in plots:\n",
    "        if plot['plot_type'] == 'line':\n",
    "            plot = ax.plot(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "        elif plot['plot_type'] == 'scatter':\n",
    "            plot = ax.scatter(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "    \n",
    "    _ = ax.set_ylim(ylim)\n",
    "    # add the color chroms\n",
    "    _ = make_color_chroms(ax, chrom_df, color_chroms, ylimits=ax.get_ylim(), colormap=colormap)\n",
    "    \n",
    "    \n",
    "    ######## LABELS ###################\n",
    "    # set the axis labels\n",
    "    _ = ax.set_xlabel('genomic coords', fontsize=1.25*label_size);\n",
    "    # quick fix for one y-label\n",
    "    _ = ax.set_ylabel(' / '.join([plot['title'] for plot in plots]), fontsize=1.25*label_size)\n",
    "    \n",
    "    ######## CHROM LABELS #############\n",
    "    add_chrom_labels(ax, chrom_df, ax.get_ylim())\n",
    "    \n",
    "    ####### X-AXIS ####################\n",
    "    # set major ticks and grid for chrom\n",
    "    \n",
    "    ax = set_ticks(ax, df, chrom_df, label_size=label_size)\n",
    "    \n",
    "    # return fig and ax for further plotting and return edited dataframe\n",
    "    return fig, ax, df, chrom_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-28T10:11:24.678525Z",
     "start_time": "2020-10-28T10:11:23.484504Z"
    }
   },
   "outputs": [],
   "source": [
    "log2 = dict(\n",
    "        title='log2ratio',\n",
    "        plot_type='scatter',   # ['line', 'scatter']\n",
    "        data='log2ratio',\n",
    "        plot_args=dict(\n",
    "            linewidth=0.3,\n",
    "            color='black',\n",
    "            s=0.2,\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "log2mean = dict(\n",
    "        title='rollinglog2ratio',\n",
    "        plot_type='line',   # ['line', 'scatter']\n",
    "        data='log2ratiomean',\n",
    "        plot_args=dict(\n",
    "            linewidth=1,\n",
    "            color='yellow',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "chroms = ['chr5', 'chr7','chr8', 'chr11', 'chr17']\n",
    "\n",
    "fig_params = dict(\n",
    "    figsize=(24,8),\n",
    "    colormap='coolwarm_r',\n",
    "    color_chroms=True,\n",
    "    ylim=(-1.5,2.5),\n",
    "    label_size=15\n",
    ")\n",
    "\n",
    "fig, ax, df, chrom_df = plot_genomic(cov_df, plots=[log2, log2mean], chroms=chroms, region='', **fig_params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# visualize heteroSNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-28T09:09:37.398050Z",
     "start_time": "2020-10-28T09:09:36.971833Z"
    }
   },
   "outputs": [],
   "source": [
    "sample = \"01_A\"\n",
    "snp_df = pd.read_csv(os.path.join(output_path, f'rollingCNV/{sample}.snpcov.csv'), sep='\\t')\n",
    "snp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### for hetSNP visualization:\n",
    "    + covData is normalized to range 1 and pushed up 1.5\n",
    "    + cov_vis is specified by cov_plots\n",
    "    + snp_vis in specified by snp_plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-28T10:14:25.653380Z",
     "start_time": "2020-10-28T10:14:25.641577Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_snp(df, snp_plots=[], cov_plots=[], chroms='all', cov_offset=.25, cov_height=0.5, color_chroms=True, colormap='coolwarm_r', region='', label_size=12, figsize=(20,4), ylim=(-1,1)):\n",
    "    \n",
    "    MAXLOG2RATIO = 2.5\n",
    "    #### DATA MANGELING ##########\n",
    "    # get cols for rearranging\n",
    "    org_cols = list(df.columns)\n",
    "    \n",
    "    # sort the df\n",
    "    df = sort_df(df)\n",
    "    # reduce the df to the selected chromosomes\n",
    "    if region:\n",
    "        chrom, start, end = extract_pos(region)\n",
    "        df = df.query('Chr == @chrom and @start <= Pos <= @end')\n",
    "    elif chroms != 'all':\n",
    "        df = df.query('Chr in @chroms')\n",
    "\n",
    "    # get the chrom_df for collapsing the \n",
    "    chrom_df = get_chrom_df(df)\n",
    "    \n",
    "    df = df.merge(chrom_df.loc[:, 'dif'], on='Chr')\n",
    "    df['PlotPos'] = df['FullExonPos'] - df['dif']\n",
    "    \n",
    "    # rearrange the df as return value\n",
    "    new_cols = org_cols[:4] + ['PlotPos'] + org_cols[4:]\n",
    "    df = df.loc[:, new_cols]\n",
    "    \n",
    "    #########################\n",
    "    ######## PLOTTING #######\n",
    "    # plot the figure\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "    # set the x-axis limits\n",
    "    _ = ax.set_xlim(0, df['PlotPos'].max())\n",
    "    \n",
    "    \n",
    "    ######## PLOT COV Data\n",
    "    \n",
    "    if len(cov_plots):\n",
    "        scale_factor = cov_height / (MAXLOG2RATIO + 1)\n",
    "        offset = 1 + scale_factor + cov_offset\n",
    "\n",
    "        ylim = (ylim[0], ylim[1] + cov_offset + cov_height)\n",
    "        \n",
    "        for plot in cov_plots:\n",
    "            # normalize the coverage data:\n",
    "            # 2.5 is the approx max log2ratio (LOH to 8N)\n",
    "\n",
    "            df[plot['data']] = df[plot['data']] * scale_factor + offset\n",
    "            if plot['plot_type'] == 'line':\n",
    "                plot = ax.plot(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "            elif plot['plot_type'] == 'scatter':\n",
    "                plot = ax.scatter(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "    \n",
    "    ######## plot the SNP graphs #######\n",
    "    for plot in snp_plots:\n",
    "        if plot['plot_type'] == 'line':\n",
    "            plot = ax.plot(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "        elif plot['plot_type'] == 'scatter':\n",
    "            plot = ax.scatter(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "    \n",
    "    _ = ax.set_ylim(ylim)\n",
    "    # add the color chroms\n",
    "    _ = make_color_chroms(ax, chrom_df, color_chroms, ylimits=ax.get_ylim(), colormap=colormap)\n",
    "    \n",
    "    \n",
    "    ######## LABELS ###################\n",
    "    # set the axis labels\n",
    "    _ = ax.set_xlabel('genomic coords', fontsize=1.25*label_size);\n",
    "    # quick fix for one y-label\n",
    "    _ = ax.set_ylabel(' / '.join([plot['title'] for plot in plots]), fontsize=1.25*label_size)\n",
    "    \n",
    "    ######## CHROM LABELS #############\n",
    "    add_chrom_labels(ax, chrom_df, ax.get_ylim())\n",
    "    \n",
    "    ####### X-AXIS ####################\n",
    "    # set major ticks and grid for chrom\n",
    "    \n",
    "    ax = set_ticks(ax, df, chrom_df, label_size=label_size)\n",
    "    \n",
    "    # return fig and ax for further plotting and return edited dataframe\n",
    "    return fig, ax, df, chrom_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-28T10:14:47.872654Z",
     "start_time": "2020-10-28T10:14:47.214561Z"
    }
   },
   "outputs": [],
   "source": [
    "log2 = dict(\n",
    "        title='log2ratio',\n",
    "        plot_type='scatter',   # ['line', 'scatter']\n",
    "        data='log2ratio',\n",
    "        plot_args=dict(\n",
    "            linewidth=0.3,\n",
    "            color='black',\n",
    "            s=0.2,\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "log2mean = dict(\n",
    "        title='rollinglog2ratio',\n",
    "        plot_type='line',   # ['line', 'scatter']\n",
    "        data='log2ratiomean',\n",
    "        plot_args=dict(\n",
    "            linewidth=1,\n",
    "            color='yellow',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "vaf = dict(\n",
    "        title='VAF',\n",
    "        plot_type='scatter',  # ['line', 'scatter']\n",
    "        data='VAF',\n",
    "        plot_args=dict(\n",
    "            s=.2,\n",
    "            color='black',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "fig_params = dict(\n",
    "    figsize=(24,8),\n",
    "    colormap='coolwarm_r',\n",
    "    color_chroms=True,\n",
    "    ylim=(0,1),\n",
    "    cov_offset=.1,  # how much log2ratio=0 is shifted above SNP-data\n",
    "    cov_height=.5,\n",
    "    label_size=13\n",
    ")\n",
    "\n",
    "fig, ax, df, chrom_df = plot_snp(snp_df, snp_plots=[vaf], cov_plots=[log2, log2mean], chroms=chroms, region='', **fig_params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## visualize clusterings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### load example clustered file from DBscan clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T16:13:52.952933Z",
     "start_time": "2020-10-30T16:13:52.855706Z"
    }
   },
   "outputs": [],
   "source": [
    "sample = \"01_A\"\n",
    "df = pd.read_csv(os.path.join(output_path, f'cluster/{sample}.dbscan.csv'), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T16:13:56.217723Z",
     "start_time": "2020-10-30T16:13:56.175961Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Pos</th>\n",
       "      <th>FullExonPos</th>\n",
       "      <th>ExonPos</th>\n",
       "      <th>log2ratio</th>\n",
       "      <th>log2ratiomean</th>\n",
       "      <th>log2ratiomeanDiff</th>\n",
       "      <th>Depth</th>\n",
       "      <th>EBscore</th>\n",
       "      <th>VAF</th>\n",
       "      <th>absVAF</th>\n",
       "      <th>deltaVAF</th>\n",
       "      <th>absVAFmean</th>\n",
       "      <th>absVAFmeanDiff</th>\n",
       "      <th>VAFstd</th>\n",
       "      <th>VAFstdDiff</th>\n",
       "      <th>deltaVAFvar</th>\n",
       "      <th>deltaVAFvarDiff</th>\n",
       "      <th>deltaVAFstd</th>\n",
       "      <th>deltaVAFstdDiff</th>\n",
       "      <th>dbscan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1054900</td>\n",
       "      <td>37368</td>\n",
       "      <td>37368</td>\n",
       "      <td>-0.069236</td>\n",
       "      <td>0.346636</td>\n",
       "      <td>3.945632e-03</td>\n",
       "      <td>147.0</td>\n",
       "      <td>0.044</td>\n",
       "      <td>0.421769</td>\n",
       "      <td>0.156462</td>\n",
       "      <td>0.056079</td>\n",
       "      <td>0.310983</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.178631</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.042368</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.205834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1055000</td>\n",
       "      <td>37468</td>\n",
       "      <td>37468</td>\n",
       "      <td>-0.066106</td>\n",
       "      <td>0.346110</td>\n",
       "      <td>3.457801e-03</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.919</td>\n",
       "      <td>0.489796</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>0.068027</td>\n",
       "      <td>0.310983</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.178631</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.042368</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.205834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1055037</td>\n",
       "      <td>37505</td>\n",
       "      <td>37505</td>\n",
       "      <td>-0.031916</td>\n",
       "      <td>0.346188</td>\n",
       "      <td>3.524551e-03</td>\n",
       "      <td>86.0</td>\n",
       "      <td>0.840</td>\n",
       "      <td>0.476744</td>\n",
       "      <td>0.046512</td>\n",
       "      <td>0.013052</td>\n",
       "      <td>0.310983</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.178631</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.042368</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.205834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1071823</td>\n",
       "      <td>37706</td>\n",
       "      <td>37706</td>\n",
       "      <td>0.337613</td>\n",
       "      <td>0.346937</td>\n",
       "      <td>3.793557e-03</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.532110</td>\n",
       "      <td>0.064220</td>\n",
       "      <td>0.055366</td>\n",
       "      <td>0.310983</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.178631</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.042368</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.205834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1291377</td>\n",
       "      <td>55139</td>\n",
       "      <td>55139</td>\n",
       "      <td>0.242629</td>\n",
       "      <td>0.247084</td>\n",
       "      <td>8.714793e-10</td>\n",
       "      <td>182.0</td>\n",
       "      <td>1.818</td>\n",
       "      <td>0.445055</td>\n",
       "      <td>0.109890</td>\n",
       "      <td>0.087055</td>\n",
       "      <td>0.310983</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.178631</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.042368</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.205834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24009</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156023260</td>\n",
       "      <td>49469650</td>\n",
       "      <td>1835187</td>\n",
       "      <td>-0.667568</td>\n",
       "      <td>0.212712</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.950</td>\n",
       "      <td>0.880000</td>\n",
       "      <td>0.760000</td>\n",
       "      <td>0.023226</td>\n",
       "      <td>0.155058</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.141056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015260</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.123530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24010</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156023266</td>\n",
       "      <td>49469656</td>\n",
       "      <td>1835193</td>\n",
       "      <td>-0.655721</td>\n",
       "      <td>0.212684</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.930</td>\n",
       "      <td>0.880000</td>\n",
       "      <td>0.760000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.155058</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.141056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015260</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.123530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24011</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156024203</td>\n",
       "      <td>49469886</td>\n",
       "      <td>1835423</td>\n",
       "      <td>0.148033</td>\n",
       "      <td>0.211628</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.392</td>\n",
       "      <td>0.132075</td>\n",
       "      <td>0.735850</td>\n",
       "      <td>0.747925</td>\n",
       "      <td>0.155058</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.141056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015260</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.123530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24012</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156024239</td>\n",
       "      <td>49469922</td>\n",
       "      <td>1835459</td>\n",
       "      <td>0.160322</td>\n",
       "      <td>0.211463</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.267857</td>\n",
       "      <td>0.464286</td>\n",
       "      <td>0.135782</td>\n",
       "      <td>0.155058</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.141056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015260</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.123530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24013</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156024251</td>\n",
       "      <td>49469934</td>\n",
       "      <td>1835471</td>\n",
       "      <td>0.164918</td>\n",
       "      <td>0.211408</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>54.0</td>\n",
       "      <td>0.304</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.156746</td>\n",
       "      <td>0.155058</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.141056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015260</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.123530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24014 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr        Pos  FullExonPos  ExonPos  log2ratio  log2ratiomean  \\\n",
       "0      chr1    1054900        37368    37368  -0.069236       0.346636   \n",
       "1      chr1    1055000        37468    37468  -0.066106       0.346110   \n",
       "2      chr1    1055037        37505    37505  -0.031916       0.346188   \n",
       "3      chr1    1071823        37706    37706   0.337613       0.346937   \n",
       "4      chr1    1291377        55139    55139   0.242629       0.247084   \n",
       "...     ...        ...          ...      ...        ...            ...   \n",
       "24009  chrX  156023260     49469650  1835187  -0.667568       0.212712   \n",
       "24010  chrX  156023266     49469656  1835193  -0.655721       0.212684   \n",
       "24011  chrX  156024203     49469886  1835423   0.148033       0.211628   \n",
       "24012  chrX  156024239     49469922  1835459   0.160322       0.211463   \n",
       "24013  chrX  156024251     49469934  1835471   0.164918       0.211408   \n",
       "\n",
       "       log2ratiomeanDiff  Depth  EBscore       VAF    absVAF  deltaVAF  \\\n",
       "0           3.945632e-03  147.0    0.044  0.421769  0.156462  0.056079   \n",
       "1           3.457801e-03   98.0    0.919  0.489796  0.020408  0.068027   \n",
       "2           3.524551e-03   86.0    0.840  0.476744  0.046512  0.013052   \n",
       "3           3.793557e-03  109.0    0.007  0.532110  0.064220  0.055366   \n",
       "4           8.714793e-10  182.0    1.818  0.445055  0.109890  0.087055   \n",
       "...                  ...    ...      ...       ...       ...       ...   \n",
       "24009       0.000000e+00   25.0    0.950  0.880000  0.760000  0.023226   \n",
       "24010       0.000000e+00   25.0    0.930  0.880000  0.760000  0.000000   \n",
       "24011       0.000000e+00   53.0    0.392  0.132075  0.735850  0.747925   \n",
       "24012       0.000000e+00   56.0    0.203  0.267857  0.464286  0.135782   \n",
       "24013       0.000000e+00   54.0    0.304  0.111111  0.777778  0.156746   \n",
       "\n",
       "       absVAFmean  absVAFmeanDiff    VAFstd  VAFstdDiff  deltaVAFvar  \\\n",
       "0        0.310983             0.0  0.178631         0.0     0.042368   \n",
       "1        0.310983             0.0  0.178631         0.0     0.042368   \n",
       "2        0.310983             0.0  0.178631         0.0     0.042368   \n",
       "3        0.310983             0.0  0.178631         0.0     0.042368   \n",
       "4        0.310983             0.0  0.178631         0.0     0.042368   \n",
       "...           ...             ...       ...         ...          ...   \n",
       "24009    0.155058             0.0  0.141056         0.0     0.015260   \n",
       "24010    0.155058             0.0  0.141056         0.0     0.015260   \n",
       "24011    0.155058             0.0  0.141056         0.0     0.015260   \n",
       "24012    0.155058             0.0  0.141056         0.0     0.015260   \n",
       "24013    0.155058             0.0  0.141056         0.0     0.015260   \n",
       "\n",
       "       deltaVAFvarDiff  deltaVAFstd  deltaVAFstdDiff  dbscan  \n",
       "0                  0.0     0.205834              0.0       0  \n",
       "1                  0.0     0.205834              0.0       0  \n",
       "2                  0.0     0.205834              0.0       0  \n",
       "3                  0.0     0.205834              0.0       0  \n",
       "4                  0.0     0.205834              0.0       0  \n",
       "...                ...          ...              ...     ...  \n",
       "24009              0.0     0.123530              0.0       0  \n",
       "24010              0.0     0.123530              0.0       0  \n",
       "24011              0.0     0.123530              0.0       0  \n",
       "24012              0.0     0.123530              0.0       0  \n",
       "24013              0.0     0.123530              0.0       0  \n",
       "\n",
       "[24014 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
