{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T05:40:30.163058Z",
     "start_time": "2021-06-06T05:40:30.148911Z"
    }
   },
   "outputs": [],
   "source": [
    "# get the code\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "# use seaborn plotting defaults\n",
    "import seaborn as sns; sns.set()\n",
    "from matplotlib.collections import PatchCollection\n",
    "from matplotlib.patches import Rectangle\n",
    "\n",
    "# import package functions\n",
    "sys.path.append('../code')\n",
    "from script_utils_CNV import get_CNVconfig, show_output\n",
    "from rollingCNV import interpolate, one_col_rolling, llh, rolling_data\n",
    "from rollingCov import rolling_coverage\n",
    "from combineCNV import filter_snp\n",
    "from plot import plot_cov, plot_snp\n",
    "\n",
    "######## need to find the chained assignment!!!\n",
    "pd.set_option('mode.chained_assignment', None)\n",
    "\n",
    "# HOME\n",
    "home = '/Users/mahtin'\n",
    "home = '/Users/martinscience'\n",
    "\n",
    "\n",
    "# standard paths\n",
    "static = os.path.join(home, \"Dropbox/Icke/Work/static\")\n",
    "tooldata = os.path.join(home, \"Dropbox/Icke/Work/somVar/tooldata\")\n",
    "testdata = os.path.join(home,\"Dropbox/Icke/Work/somVar/testdata\")\n",
    "PON_path = os.path.join(static, \"PON/HAEv7_hg38_NovaSeq\")\n",
    "\n",
    "\n",
    "cnvdata = os.path.join(tooldata, \"myCNVdata\")\n",
    "output_path = os.path.join(cnvdata, \"output\")\n",
    "plot_path = os.path.join(cnvdata, \"plot\")\n",
    "fig_path = os.path.join(cnvdata, \"figures\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## visualize coverage data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T06:31:29.800283Z",
     "start_time": "2021-06-06T06:31:19.805754Z"
    }
   },
   "outputs": [],
   "source": [
    "sample = \"03_A-B\"\n",
    "cnv_df = pd.read_csv(os.path.join(output_path, f\"CNV/{sample}.cnv.full.gz\"), sep=\"\\t\", compression=\"gzip\")\n",
    "rsnp_df = pd.read_csv(os.path.join(output_path, f\"CNV/{sample}.cnv.snp.gz\"), sep=\"\\t\", compression=\"gzip\")\n",
    "cnv_df[:3]\n",
    "rsnp_df[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:11:09.339886Z",
     "start_time": "2020-10-30T20:11:08.477294Z"
    }
   },
   "source": [
    "### load helpers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T08:54:37.396822Z",
     "start_time": "2021-06-06T08:54:37.372794Z"
    }
   },
   "outputs": [],
   "source": [
    "def sort_df(df):\n",
    "    '''\n",
    "    helper for sorting dfs for chromosomes\n",
    "    '''\n",
    "    df2 = df.copy()\n",
    "    # make Chr column categorical for sorting .. and sort\n",
    "    chrom_list = [f\"chr{i+1}\" for i in range(22)] + ['chrX']\n",
    "    df2['Chr'] = pd.Categorical(df2['Chr'], chrom_list)\n",
    "    return df2.sort_values(['Chr', 'FullExonPos'])\n",
    "\n",
    "\n",
    "def get_chrom_df(df):\n",
    "\n",
    "    # dropna is neccessary because grouping on categorical returns all categories\n",
    "    chrom_df = df.groupby('Chr')['FullExonPos'].agg(['mean', 'min', 'max']).dropna()\n",
    "    cols = list(chrom_df.columns)\n",
    "    chrom_df['sum'] = chrom_df['max'] - chrom_df['min']\n",
    "    chrom_df['cummin'] = chrom_df['sum'].cumsum()\n",
    "    chrom_df['dif'] = (chrom_df['max'] - chrom_df['cummin']).astype(int)\n",
    "    for col in cols:\n",
    "        chrom_df[col] = (chrom_df[col] - chrom_df['dif']).astype(int)\n",
    "    cols.append('dif')\n",
    "    return chrom_df.loc[:, cols]\n",
    "\n",
    "\n",
    "def make_color_chroms(\n",
    "    ax, chrom_df, color_chroms, ylimits=(-10, 10), colormap=\"coolwarm_r\"\n",
    "):\n",
    "\n",
    "    # set the cmap from provided argument\n",
    "    cmap = plt.cm.get_cmap(colormap, 23)\n",
    "    chrom_list = [f\"chr{i+1}\" for i in range(22)] + [\"chrX\"]\n",
    "    # build the rects\n",
    "    rects = []\n",
    "    # set the height and ymin beyond the ylimits so borders are not seen\n",
    "    ymin = ylimits[0] * 1.1\n",
    "    height = (ylimits[1] - ymin) * 1.1\n",
    "\n",
    "    for chrom in chrom_list:\n",
    "        if chrom in chrom_df.index:\n",
    "            row = chrom_df.loc[chrom]\n",
    "            rect = Rectangle(\n",
    "                (row[\"min\"], ymin), width=row[\"max\"] - row[\"min\"], height=height\n",
    "            )\n",
    "            rects.append(rect)\n",
    "        else:\n",
    "            rect = Rectangle((0, 1), width=0.1, height=height)\n",
    "            rects.append(rect)\n",
    "    if color_chroms:\n",
    "        rect_kwargs = dict(alpha=0.4, ec=\"none\")\n",
    "    else:\n",
    "        rect_kwargs = dict(alpha=1, fc=\"none\", ec=\"darkgray\", lw=1, ls=\"-\")\n",
    "    # set the rectangle collection with colormap\n",
    "    rect_collection = PatchCollection(rects, cmap=cmap, **rect_kwargs)\n",
    "    # set the index for the color map from chrom integers\n",
    "    rect_collection.set_array(\n",
    "        chrom_df.index.str.replace(\"chr\", \"\").str.replace(\"X\", \"23\").astype(int)\n",
    "    )\n",
    "\n",
    "    # setting clim allows fixing the color for individual chroms\n",
    "    # https://stackoverflow.com/questions/6028675/setting-color-range-in-matplotlib-patchcollection\n",
    "    rect_collection.set_clim([0, 23])\n",
    "    return ax.add_collection(rect_collection)\n",
    "\n",
    "\n",
    "def add_chrom_labels(ax, chrom_df, ylimits=(-10,10)):\n",
    "    \n",
    "    # YOFFSET is the upper-relative y-position \n",
    "    YOFFSET = 1\n",
    "    # get the min_chrom_fraction from minimum chrom_size / whole stretch\n",
    "    min_chrom_frac = (chrom_df['max'] - chrom_df['min']).min() / chrom_df['max'].max()\n",
    "    chrom_size = min(25, max(15, 200*min_chrom_frac))\n",
    "    style = dict(size=chrom_size, color='#2f3832')\n",
    "    # set the height and ymin beyond the ylimits so borders are not seen\n",
    "    ypos = ylimits[0] + YOFFSET * (ylimits[1] - ylimits[0])\n",
    "    for chrom, row in chrom_df.iterrows():\n",
    "        if (len(chrom_df.index) > 12):\n",
    "            chrom = chrom.replace('chr', '')\n",
    "        ax.text(row['mean'], ypos, chrom, ha='center', **style)\n",
    "    \n",
    "\n",
    "def make_nice(position):\n",
    "    '''\n",
    "    takes position and returns closest multiple of 1, 2, 5 or 10\n",
    "    '''\n",
    "    # set nice values\n",
    "    nice_positions = np.array([1,2,2.5,5,10])\n",
    "    # get the 10s\n",
    "    power10 = np.power(10, np.floor(np.log10(position)))\n",
    "    # reduce to value between 1 and 10\n",
    "    num = position / power10\n",
    "    # find the closest nice position\n",
    "    base = nice_positions[np.argmin(np.abs(nice_positions / num - 1))]\n",
    "    return base * power10\n",
    "\n",
    "def get_tick_pos(tick_dist, chrom_df):\n",
    "    '''\n",
    "    return from chrom_df the evenly-spread (tick_dist) positions per chrom\n",
    "    '''\n",
    "    return [pos for _, row in chrom_df.iterrows() for pos in range(row['min'] + tick_dist, row['max'], tick_dist)]\n",
    "\n",
    "\n",
    "def str_pos(pos, df, precision=1):\n",
    "    \"\"\"\n",
    "    returns string representation of base position\n",
    "    on genomic coords\n",
    "    \"\"\"\n",
    "    pos = df.iloc[np.argmin(np.abs(df[\"PlotPos\"] - pos))][\"Pos\"]\n",
    "    # get the closest base power\n",
    "    power10 = int(np.round(np.log10(pos) / 3) * 3)\n",
    "    # get the base fraction\n",
    "    base = pos / np.power(10, power10)\n",
    "    if power10 == 9:\n",
    "        base = base * 1000\n",
    "        power10 = 6\n",
    "    if power10 == 6:\n",
    "        suff = \"Mb\"\n",
    "    elif power10 == 3:\n",
    "        suff = \"kb\"\n",
    "    base = re.sub(r\"\\.0$\", \"\", str(round(base, precision)))\n",
    "    if power10 == 0:\n",
    "        suff = \"b\"\n",
    "        base = int(base)\n",
    "    # only print the suff if print_suff\n",
    "    return f\"{base}{suff}\"\n",
    "\n",
    "\n",
    "def get_precision(pos_list):\n",
    "    \"\"\"\n",
    "    get the major tick precision from the range of chrom values\n",
    "    \"\"\"\n",
    "    # get the range of positions\n",
    "    prange = max(pos_list) - min(pos_list)\n",
    "    # get the closest 10base\n",
    "    power10 = int(np.round(np.log10(prange)))\n",
    "    precision = max(7 - power10, 0)\n",
    "    return precision\n",
    "\n",
    "\n",
    "def set_ticks(ax, df, chrom_df, ticks=20, label_size=12):\n",
    "    '''\n",
    "    for a given tick number, set nicely spread ticks\n",
    "    '''\n",
    "    \n",
    "    ## determine optimale tick distance\n",
    "    # get the chrom_number\n",
    "    chrom_count = len(chrom_df.index)\n",
    "    # get the number of bases\n",
    "    stretch = chrom_df['max'][-1]\n",
    "    # set the number of desired ticks\n",
    "    major_tick_dist = int(stretch / (ticks + 1))\n",
    "    minor_tick_dist = int(stretch / ((ticks * 2) + 1))\n",
    "    \n",
    "    # feed tick distance into chrom_df to get chrom-specific coords\n",
    "    major_pos = get_tick_pos(major_tick_dist, chrom_df)\n",
    "    minor_pos = [pos - minor_tick_dist for pos in major_pos]\n",
    "    \n",
    "    ax.xaxis.set_major_locator(plt.FixedLocator(major_pos))\n",
    "    # only print the genomic coords below a certain base total\n",
    "    if stretch < 2e7:\n",
    "        precision = get_precision(major_pos)\n",
    "        major_labels = [\n",
    "            str_pos(pos, df, precision=precision)  ###############\n",
    "            for pos in major_pos\n",
    "        ]\n",
    "        ax.xaxis.set_major_formatter(plt.FixedFormatter(major_labels))\n",
    "        # set the axis labels\n",
    "        ax.set_xlabel(\"genomic coords\", fontsize=1.25 * label_size)\n",
    "    else:\n",
    "        ax.xaxis.set_major_formatter(plt.NullFormatter())\n",
    "    ax.xaxis.set_minor_locator(plt.FixedLocator(minor_pos))\n",
    "    ax.xaxis.grid(which='major', linestyle='-', linewidth=2)\n",
    "    ax.xaxis.grid(which='minor', linestyle='--', linewidth=1)\n",
    "    ax.xaxis.set_tick_params(which='major', length=20, labelsize=label_size)\n",
    "    ax.yaxis.set_tick_params(which='major', length=20, labelsize=label_size)\n",
    "    # set the tick labels\n",
    "    for tick in ax.xaxis.get_majorticklabels():\n",
    "        tick.set_verticalalignment(\"bottom\")\n",
    "    return ax\n",
    "\n",
    "\n",
    "def extract_pos(region):\n",
    "    def convert(pos):\n",
    "        if pos.endswith(\"Mb\"):\n",
    "            pos = int(float(pos.replace(\"Mb\", \"\")) * 1e6)\n",
    "        elif pos.endswith(\"kb\"):\n",
    "            pos = int(float(pos.replace(\"kb\", \"\")) * 1000)\n",
    "        else:\n",
    "            pos = int(pos)\n",
    "        return pos\n",
    "\n",
    "    split = region.split(\":\")\n",
    "    chrom = split[0]\n",
    "\n",
    "    # if start and are used\n",
    "    if len(split) > 1 and \"-\" in split[1]:\n",
    "        se = split[1].split(\"-\")\n",
    "        start = convert(se[0])\n",
    "        end = convert(se[1])\n",
    "    else:\n",
    "        start = 0\n",
    "        end = 1e10\n",
    "    return chrom, start, end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## coverage plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T08:48:03.506825Z",
     "start_time": "2021-06-06T08:48:03.499251Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_cov(\n",
    "    df,\n",
    "    plots,\n",
    "    chroms=\"all\",\n",
    "    color_chroms=True,\n",
    "    colormap=\"coolwarm_r\",\n",
    "    region=\"\",\n",
    "    figsize=(20, 4),\n",
    "    ylim=(-1, 1),\n",
    "    label_size=12,\n",
    "):\n",
    "\n",
    "    # ### DATA MANGLING ##########\n",
    "    # get cols for rearranging\n",
    "    org_cols = list(df.columns)\n",
    "\n",
    "    # sort the df\n",
    "    df = sort_df(df)\n",
    "    # reduce the df to the selected chromosomes\n",
    "    if region:\n",
    "        chrom, start, end = extract_pos(region)\n",
    "        df = df.query(\"Chr == @chrom and @start <= Pos <= @end\")\n",
    "    elif chroms != \"all\":\n",
    "        df = df.query(\"Chr in @chroms\")\n",
    "\n",
    "    # get the chrom_df for collapsing the\n",
    "    chrom_df = get_chrom_df(df)\n",
    "\n",
    "    df = df.merge(chrom_df.loc[:, \"dif\"], on=\"Chr\")\n",
    "    df[\"PlotPos\"] = df[\"FullExonPos\"] - df[\"dif\"]\n",
    "\n",
    "    # rearrange the df as return value\n",
    "    new_cols = org_cols[:4] + [\"PlotPos\"] + org_cols[4:]\n",
    "    df = df.loc[:, new_cols]\n",
    "\n",
    "    # ####### PLOTTING #######\n",
    "    # plot the figure\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "    # set the x-axis limits\n",
    "    _ = ax.set_xlim(0, df[\"PlotPos\"].max())\n",
    "\n",
    "    # plot the graphs #######\n",
    "    for plot in plots:\n",
    "        if plot[\"plot_type\"] == \"line\":\n",
    "            plot = ax.plot(df[\"PlotPos\"], df[plot[\"data\"]], **plot[\"plot_args\"])\n",
    "        elif plot[\"plot_type\"] == \"scatter\":\n",
    "            plot = ax.scatter(df[\"PlotPos\"], df[plot[\"data\"]], **plot[\"plot_args\"])\n",
    "\n",
    "    _ = ax.set_ylim(ylim)\n",
    "    # add the color chroms\n",
    "    _ = make_color_chroms(\n",
    "        ax, chrom_df, color_chroms, ylimits=ax.get_ylim(), colormap=colormap\n",
    "    )\n",
    "\n",
    "    # ####### LABELS ###################\n",
    "\n",
    "    # quick fix for one y-label\n",
    "    _ = ax.set_ylabel(\n",
    "        \" / \".join([plot[\"title\"] for plot in plots]), fontsize=1.25 * label_size\n",
    "    )\n",
    "\n",
    "    # ####### CHROM LABELS #############\n",
    "    add_chrom_labels(ax, chrom_df, ax.get_ylim())\n",
    "\n",
    "    # set chrom borders\n",
    "    for m in chrom_df[\"min\"][1:]:\n",
    "        _ = ax.axvline(x=m, c=\"k\", lw=0.5, alpha=0.5, ls=\"-\")\n",
    "    # ###### X-AXIS ####################\n",
    "    # set major ticks and grid for chrom\n",
    "    ax = set_ticks(ax, df, chrom_df, label_size=label_size)\n",
    "\n",
    "    # return fig and ax for further plotting and return edited dataframe\n",
    "    return fig, ax, df, chrom_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T08:54:46.351307Z",
     "start_time": "2021-06-06T08:54:41.610093Z"
    }
   },
   "outputs": [],
   "source": [
    "log2 = dict(\n",
    "        title='L2R',\n",
    "        plot_type='scatter',   # ['line', 'scatter']\n",
    "        data='log2ratio2',\n",
    "        plot_args=dict(\n",
    "            linewidth=0.3,\n",
    "            color='black',\n",
    "            s=0.8,\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "log2mean = dict(\n",
    "        title='L2R_mean',\n",
    "        plot_type='line',   # ['line', 'scatter']\n",
    "        data='log2ratio2_mean',\n",
    "        plot_args=dict(\n",
    "            linewidth=1,\n",
    "            color='yellow',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "chroms = ['chr5', 'chr7','chr8', 'chr11', 'chr17']\n",
    "\n",
    "fig_params = dict(\n",
    "    figsize=(34,4),\n",
    "    colormap='coolwarm_r',\n",
    "    color_chroms=True,\n",
    "    ylim=(-1.5,2.5),\n",
    "    label_size=15\n",
    ")\n",
    "chroms = ['chr3', 'chr4', 'chr5', 'chr6','chr7', 'chr9', 'chr12', 'chr17']\n",
    "r20 = 'chr20:20.7Mb-34.5Mb'\n",
    "r7 = 'chr7:95.8Mb-111Mb'\n",
    "r17 = 'chr17:18.2Mb-25Mb'\n",
    "fig, ax, df, chrom_df = plot_cov(cnv_df, plots=[log2, log2mean], chroms=\"all\", region= 'chr7:95.8Mb-110Mb', **fig_params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T08:36:36.312234Z",
     "start_time": "2021-06-06T08:36:28.053206Z"
    }
   },
   "outputs": [],
   "source": [
    "fig, ax, df, chrom_df = plot_cov(cnv_df.query('log2ratio2 == log2ratio2'), plots=[log2, log2mean], chroms=\"all\", region='', **fig_params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# visualize heteroSNP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:11:23.058284Z",
     "start_time": "2020-10-30T20:11:22.639094Z"
    }
   },
   "source": [
    "sample = \"03_A-B\"\n",
    "snp_df = pd.read_csv(os.path.join(output_path, f\"snp/{sample}.snp.gz\"), sep=\"\\t\", compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T06:29:22.152133Z",
     "start_time": "2021-06-06T06:29:22.098776Z"
    }
   },
   "outputs": [],
   "source": [
    "snp_df = cnv_df.query(\"VAF2 == VAF2\")\n",
    "snp_df[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### for hetSNP visualization:\n",
    "    + snp_vis in specified by snp_plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T08:55:30.726075Z",
     "start_time": "2021-06-06T08:55:30.716467Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_snp(\n",
    "    df,\n",
    "    plots=[],\n",
    "    chroms=\"all\",\n",
    "    color_chroms=True,\n",
    "    colormap=\"coolwarm_r\",\n",
    "    region=\"\",\n",
    "    label_size=12,\n",
    "    figsize=(20, 4),\n",
    "    ylim=(-1, 1),\n",
    "):\n",
    "\n",
    "    # ### DATA MANGELING ##########\n",
    "    # get cols for rearranging\n",
    "    org_cols = list(df.columns)\n",
    "\n",
    "    # sort the df\n",
    "    df = sort_df(df)\n",
    "    # reduce the df to the selected chromosomes\n",
    "    if region:\n",
    "        chrom, start, end = extract_pos(region)\n",
    "        df = df.query(\"Chr == @chrom and @start <= Pos <= @end\")\n",
    "    elif chroms != \"all\":\n",
    "        df = df.query(\"Chr in @chroms\")\n",
    "\n",
    "    # get the chrom_df for collapsing\n",
    "    chrom_df = get_chrom_df(df)\n",
    "    df = df.merge(chrom_df.loc[:, \"dif\"], on=\"Chr\")\n",
    "    df[\"PlotPos\"] = df[\"FullExonPos\"] - df[\"dif\"]\n",
    "\n",
    "    # rearrange the df as return value\n",
    "    new_cols = org_cols[:4] + [\"PlotPos\"] + org_cols[4:]\n",
    "    df = df.loc[:, new_cols]\n",
    "\n",
    "    # ########################\n",
    "    # ####### PLOTTING #######\n",
    "    # plot the figure\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "    # set the x-axis limits\n",
    "    _ = ax.set_xlim(0, df[\"PlotPos\"].max())\n",
    "\n",
    "    # ####### plot the SNP graphs #######\n",
    "    for plot in plots:\n",
    "        if plot[\"plot_type\"] == \"line\":\n",
    "            plot = ax.plot(df[\"PlotPos\"], df[plot[\"data\"]], **plot[\"plot_args\"])\n",
    "        elif plot[\"plot_type\"] == \"scatter\":\n",
    "            # highjack plot_args with\n",
    "            pa = plot[\"plot_args\"]\n",
    "            if \"c\" in pa:\n",
    "                pa[\"c\"] = df[pa[\"c\"]]\n",
    "            if \"s\" in pa:\n",
    "                if isinstance(pa[\"s\"], str):\n",
    "                    pa[\"s\"] = df[pa[\"s\"]] * 20 + 1\n",
    "            plot = ax.scatter(df[\"PlotPos\"], df[plot[\"data\"]], **pa)\n",
    "\n",
    "    _ = ax.set_ylim(ylim)\n",
    "    # add the color chroms\n",
    "    _ = make_color_chroms(\n",
    "        ax, chrom_df, color_chroms, ylimits=ax.get_ylim(), colormap=colormap\n",
    "    )\n",
    "\n",
    "    # ####### LABELS ###################\n",
    "\n",
    "    # quick fix for one y-label\n",
    "    _ = ax.set_ylabel(\n",
    "        \" / \".join([plot[\"title\"] for plot in plots]), fontsize=1.25 * label_size\n",
    "    )\n",
    "\n",
    "    # ####### CHROM LABELS #############\n",
    "    add_chrom_labels(ax, chrom_df, ax.get_ylim())\n",
    "\n",
    "    # ###### X-AXIS ####################\n",
    "    # set major ticks and grid for chrom\n",
    "\n",
    "    ax = set_ticks(ax, df, chrom_df, label_size=label_size)\n",
    "    # set helper lines\n",
    "    #  _ = ax.axhline(y=1, c=\"k\", lw=2, ls=\"-\")\n",
    "    _ = ax.axhline(y=0.5, c=\"k\", lw=1.5, alpha=0.5, ls=\"--\")\n",
    "    \n",
    "    # set chrom borders\n",
    "    for m in chrom_df['min'][1:]:\n",
    "        _ = ax.axvline(x=m, c=\"k\", lw=0.5, alpha=0.5, ls=\"-\")\n",
    "    # return fig and ax for further plotting and return edited dataframe\n",
    "    return fig, ax, df, chrom_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T08:55:33.033270Z",
     "start_time": "2021-06-06T08:55:32.695713Z"
    }
   },
   "outputs": [],
   "source": [
    "vaf = dict(\n",
    "        title='VAF',\n",
    "        plot_type='scatter',  # ['line', 'scatter']\n",
    "        data='VAF2',\n",
    "        plot_args=dict(\n",
    "            s=2,\n",
    "            color='black',\n",
    "            alpha=1\n",
    "        )\n",
    "    )\n",
    "\n",
    "fig_params = dict(\n",
    "    figsize=(34,4),\n",
    "    colormap='coolwarm_r',\n",
    "    color_chroms=True,\n",
    "    ylim=(0,1),\n",
    "    label_size=13\n",
    ")\n",
    "\n",
    "fig, ax, df, chrom_df = plot_snp(snp_df, plots=[vaf], chroms=\"all\", region='', **fig_params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# combine SNP and coverage figs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T08:55:41.229595Z",
     "start_time": "2021-06-06T08:55:41.214787Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_CNV(\n",
    "    df,\n",
    "    snp_plots=[],\n",
    "    cov_plots=[],\n",
    "    chroms=\"all\",\n",
    "    color_chroms=True,\n",
    "    colormap=\"coolwarm_r\",\n",
    "    region=\"\",\n",
    "    label_size=12,\n",
    "    figsize=(20, 4),\n",
    "    ylims=dict(cov=(-1, 2.5), snp=(0, 1)),\n",
    "):\n",
    "\n",
    "    # ### DATA MANGELING ##########\n",
    "    # get cols for rearranging\n",
    "    org_cols = list(df.columns)\n",
    "    # sort the df\n",
    "    df = sort_df(df)\n",
    "    # reduce the df to the selected chromosomes\n",
    "    if region:\n",
    "        chrom, start, end = extract_pos(region)\n",
    "        df = df.query(\"Chr == @chrom and @start <= Pos <= @end\")\n",
    "    elif chroms != \"all\":\n",
    "        df = df.query(\"Chr in @chroms\")\n",
    "\n",
    "    # get the chrom_df for collapsing\n",
    "    chrom_df = get_chrom_df(df)\n",
    "    df = df.merge(chrom_df.loc[:, \"dif\"], on=\"Chr\")\n",
    "    df[\"PlotPos\"] = df[\"FullExonPos\"] - df[\"dif\"]\n",
    "\n",
    "    # rearrange the df as return value\n",
    "    new_cols = org_cols[:4] + [\"PlotPos\"] + org_cols[4:]\n",
    "    df = df.loc[:, new_cols]\n",
    "\n",
    "    # ########################\n",
    "    # ####### PLOTTING #######\n",
    "    # plot the figure\n",
    "    fig, axes = plt.subplots(2, figsize=figsize, gridspec_kw={\"height_ratios\": [1, 2]})\n",
    "\n",
    "    for ax in axes:\n",
    "        # set the x-axis limits\n",
    "        _ = ax.set_xlim(0, df[\"PlotPos\"].max())\n",
    "\n",
    "    # ######## plot COVERAGE\n",
    "    for plot in cov_plots:\n",
    "\n",
    "        df[plot[\"data\"]] = df[plot[\"data\"]]\n",
    "        if plot[\"plot_type\"] == \"line\":\n",
    "            _ = axes[0].plot(df[\"PlotPos\"], df[plot[\"data\"]], **plot[\"plot_args\"])\n",
    "\n",
    "        elif plot[\"plot_type\"] == \"scatter\":\n",
    "            # highjack plot_args\n",
    "            pa = plot[\"plot_args\"]\n",
    "            if \"c\" in pa:\n",
    "                pa[\"c\"] = df[pa[\"c\"]]\n",
    "            if \"s\" in pa:\n",
    "                if isinstance(pa[\"s\"], str):\n",
    "                    pa[\"s\"] = df[pa[\"s\"]] * 20 + 1\n",
    "            _ = axes[0].scatter(df[\"PlotPos\"], df[plot[\"data\"]], **pa)\n",
    "\n",
    "    _ = axes[0].set_ylim(ylims[\"cov\"])\n",
    "\n",
    "    # ####### plot the SNP graphs #######\n",
    "    for plot in snp_plots:\n",
    "        if plot[\"plot_type\"] == \"line\":\n",
    "            plot = axes[1].plot(df[\"PlotPos\"], df[plot[\"data\"]], **plot[\"plot_args\"])\n",
    "        elif plot[\"plot_type\"] == \"scatter\":\n",
    "            # highjack plot_args with\n",
    "            pa = plot[\"plot_args\"]\n",
    "            if \"c\" in pa:\n",
    "                pa[\"c\"] = df[pa[\"c\"]]\n",
    "            if \"s\" in pa:\n",
    "                if isinstance(pa[\"s\"], str):\n",
    "                    pa[\"s\"] = df[pa[\"s\"]] * 20 + 1\n",
    "            plot = axes[1].scatter(df[\"PlotPos\"], df[plot[\"data\"]], **pa)\n",
    "\n",
    "    _ = axes[1].set_ylim(ylims[\"snp\"])\n",
    "\n",
    "    # add the color chroms\n",
    "    for ax in axes:\n",
    "        _ = make_color_chroms(\n",
    "            ax, chrom_df, color_chroms, ylimits=ax.get_ylim(), colormap=colormap\n",
    "        )\n",
    "\n",
    "    # ####### LABELS ###################\n",
    "\n",
    "    # quick fix for one y-label\n",
    "    _ = axes[0].set_ylabel(\"COV [log2r]\", fontsize=1.25 * label_size)\n",
    "    _ = axes[1].set_ylabel(\"BAF\", fontsize=1.25 * label_size)\n",
    "\n",
    "    # ###### X-AXIS ####################\n",
    "    # chrom lables\n",
    "    add_chrom_labels(axes[1], chrom_df, ax.get_ylim())\n",
    "    # set major ticks and grid for chrom\n",
    "    axes[1] = set_ticks(axes[1], df, chrom_df, label_size=label_size)\n",
    "    # remove ticks for coverage plot\n",
    "    axes[0].xaxis.set_tick_params(which=\"both\", labelbottom=False)\n",
    "    # set helper lines\n",
    "    # cov_plot\n",
    "    for line_pos in [-1, 0, 1]:\n",
    "        _ = axes[0].axhline(y=line_pos, c=\"k\", lw=1.5, alpha=0.5, ls=\"--\")\n",
    "    #  VAF plot\n",
    "    _ = axes[1].axhline(y=0.5, c=\"k\", lw=1.5, alpha=0.5, ls=\"--\")\n",
    "\n",
    "    # set chrom borders\n",
    "    for m in chrom_df[\"min\"][1:]:\n",
    "        for ax in axes:\n",
    "            _ = ax.axvline(x=m, c=\"k\", lw=0.5, alpha=0.5, ls=\"-\")\n",
    "    # return fig and ax for further plotting and return edited dataframe\n",
    "    return fig, axes, df, chrom_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T08:55:46.759616Z",
     "start_time": "2021-06-06T08:55:41.961557Z"
    }
   },
   "outputs": [],
   "source": [
    "log2 = dict(\n",
    "        title='L2R',\n",
    "        plot_type='scatter',   # ['line', 'scatter']\n",
    "        data='log2ratio2',\n",
    "        plot_args=dict(\n",
    "            linewidth=0.3,\n",
    "            color='black',\n",
    "            s=1.8,\n",
    "            alpha=1\n",
    "        )\n",
    "    )\n",
    "\n",
    "log2mean = dict(\n",
    "        title='L2R_mean',\n",
    "        plot_type='line',   # ['line', 'scatter']\n",
    "        data='log2ratio2_mean',\n",
    "        plot_args=dict(\n",
    "            linewidth=1,\n",
    "            color='yellow',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "vaf = dict(\n",
    "        title='VAF',\n",
    "        plot_type='scatter',  # ['line', 'scatter']\n",
    "        data='VAF2',\n",
    "        plot_args=dict(\n",
    "            s=2,\n",
    "            color='black',\n",
    "            alpha=1\n",
    "        )\n",
    "    )\n",
    "\n",
    "fig_params = dict(\n",
    "    figsize=(34,6),\n",
    "    colormap='coolwarm_r',\n",
    "    color_chroms=True,\n",
    "    label_size=13,\n",
    "    ylims=dict(\n",
    "        cov=(-1.2,2),\n",
    "        snp=(0, 1)\n",
    "    )\n",
    ")\n",
    "chroms = ['chr3', 'chr4', 'chr5', 'chr6','chr7', 'chr9', 'chr12', 'chr17']\n",
    "r20 = 'chr20:20.7Mb-34.5Mb'\n",
    "r7 = 'chr7:95.8Mb-111Mb'\n",
    "r17 = 'chr17:18.2Mb-25Mb'\n",
    "\n",
    "fig, axes, df, chrom_df = plot_CNV(cnv_df, cov_plots=[log2, log2mean], snp_plots=[vaf], chroms=\"all\", region=\"chr20\", **fig_params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# plot for SNP matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T05:21:06.192471Z",
     "start_time": "2021-06-06T05:21:06.186302Z"
    }
   },
   "outputs": [],
   "source": [
    "def make_SNP_plot(sample, cnv_df):\n",
    "\n",
    "    df = cnv_df.loc[:, [c for c in cnv_df.columns if c.startswith(\"VAF\")]]\n",
    "    fig, ax = plt.subplots(figsize=(10, 10))\n",
    "    _ = ax.scatter(df['VAF1'], df['VAF2'], s=0.25, alpha=0.4)\n",
    "    _ = ax.set_xlabel(\"NVAF\", fontsize=20)\n",
    "    _ = ax.set_ylabel(\"TVAF\", fontsize=20)\n",
    "    \n",
    "    # calculate offRate\n",
    "    df0 = df[(df > 0.1).any(axis=1)]\n",
    "    n = len(df0.index)\n",
    "    df1 = df0[np.abs(df0[\"VAF1\"] - df0[\"VAF2\"]) > 0.25]\n",
    "    m = len(df1.index)\n",
    "    off_ratio = m / n * 100\n",
    "    _ = ax.set_title(\n",
    "        f\"{sample} |Â  Tumor vs Normal - offRate {round(off_ratio, 1)}\", fontsize=30\n",
    "    )\n",
    "    return fig, off_ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-06T05:21:07.144944Z",
     "start_time": "2021-06-06T05:21:06.760858Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'output_path' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-ea9aa69c0da7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0msample\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"03_A-B\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mcnv_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34mf\"CNV/{sample}.cnv.snp.gz\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcompression\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"gzip\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0m_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmake_SNP_plot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcnv_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'output_path' is not defined"
     ]
    }
   ],
   "source": [
    "sample = \"03_A-B\"\n",
    "cnv_df = pd.read_csv(os.path.join(output_path, f\"CNV/{sample}.cnv.snp.gz\"), sep=\"\\t\", compression=\"gzip\")\n",
    "_ = make_SNP_plot(sample, cnv_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# visualize clusterings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### load example clustered file from DBscan clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:52:21.378234Z",
     "start_time": "2020-10-30T20:52:21.250918Z"
    }
   },
   "outputs": [],
   "source": [
    "sample = \"01_A\"\n",
    "df = pd.read_csv(os.path.join(output_path, f'cluster/{sample}.dbscan.csv'), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:52:22.176859Z",
     "start_time": "2020-10-30T20:52:22.150268Z"
    }
   },
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:52:23.585512Z",
     "start_time": "2020-10-30T20:52:23.579889Z"
    }
   },
   "outputs": [],
   "source": [
    "log2 = dict(\n",
    "        title='log2ratio',\n",
    "        plot_type='scatter',   # ['line', 'scatter']\n",
    "        data='log2ratio',\n",
    "        plot_args=dict(\n",
    "            linewidth=0.3,\n",
    "            color='black',\n",
    "            s=0.2,\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "log2mean = dict(\n",
    "        title='rollinglog2ratio',\n",
    "        plot_type='line',   # ['line', 'scatter']\n",
    "        data='log2ratiomean',\n",
    "        plot_args=dict(\n",
    "            linewidth=1,\n",
    "            color='yellow',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "absvaf = dict(\n",
    "        title='absVAF',\n",
    "        plot_type='scatter',  # ['line', 'scatter']\n",
    "        data='absVAF',\n",
    "        plot_args=dict(\n",
    "            s=1,\n",
    "            c='dbscan',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "absvafmean = dict(\n",
    "        title='deltaVAFvar',\n",
    "        plot_type='line',   # ['line', 'scatter']\n",
    "        data='deltaVAFvar',\n",
    "        plot_args=dict(\n",
    "            linewidth=1,\n",
    "            color='blue',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "fig_params = dict(\n",
    "    figsize=(24,8),\n",
    "    colormap='coolwarm_r',\n",
    "    color_chroms=False,\n",
    "    ylim=(0,1),\n",
    "    cov_offset=.1,  # how much log2ratio=0 is shifted above SNP-data\n",
    "    cov_height=.5,\n",
    "    label_size=13\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### expand plot_snp functionality to showing clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:52:31.244421Z",
     "start_time": "2020-10-30T20:52:31.230296Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_snp(df, snp_plots=[], cov_plots=[], chroms='all', cov_offset=.25, cov_height=0.5, color_chroms=True, colormap='coolwarm_r', region='', label_size=12, figsize=(20,4), ylim=(-1,1)):\n",
    "    \n",
    "    MAXLOG2RATIO = 2.5\n",
    "    #### DATA MANGELING ##########\n",
    "    # get cols for rearranging\n",
    "    org_cols = list(df.columns)\n",
    "    \n",
    "    # sort the df\n",
    "    df = sort_df(df)\n",
    "    # reduce the df to the selected chromosomes\n",
    "    if region:\n",
    "        chrom, start, end = extract_pos(region)\n",
    "        df = df.query('Chr == @chrom and @start <= Pos <= @end')\n",
    "    elif chroms != 'all':\n",
    "        df = df.query('Chr in @chroms')\n",
    "\n",
    "    # get the chrom_df for collapsing the \n",
    "    chrom_df = get_chrom_df(df)\n",
    "    \n",
    "    df = df.merge(chrom_df.loc[:, 'dif'], on='Chr')\n",
    "    df['PlotPos'] = df['FullExonPos'] - df['dif']\n",
    "    \n",
    "    # rearrange the df as return value\n",
    "    new_cols = org_cols[:4] + ['PlotPos'] + org_cols[4:]\n",
    "    df = df.loc[:, new_cols]\n",
    "    \n",
    "    #########################\n",
    "    ######## PLOTTING #######\n",
    "    # plot the figure\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "    # set the x-axis limits\n",
    "    _ = ax.set_xlim(0, df['PlotPos'].max())\n",
    "    \n",
    "    \n",
    "    ######## PLOT COV Data\n",
    "    \n",
    "    if len(cov_plots):\n",
    "        scale_factor = cov_height / (MAXLOG2RATIO + 1)\n",
    "        offset = 1 + scale_factor + cov_offset\n",
    "\n",
    "        ylim = (ylim[0], ylim[1] + cov_offset + cov_height)\n",
    "        \n",
    "        for plot in cov_plots:\n",
    "            # normalize the coverage data:\n",
    "            # 2.5 is the approx max log2ratio (LOH to 8N)\n",
    "\n",
    "            df[plot['data']] = df[plot['data']] * scale_factor + offset\n",
    "            if plot['plot_type'] == 'line':\n",
    "                plot = ax.plot(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "            elif plot['plot_type'] == 'scatter':\n",
    "                # highjack plot_args\n",
    "                pa = plot['plot_args']\n",
    "                if \"c\" in pa:\n",
    "                    pa['c'] = df[pa['c']]\n",
    "                if \"s\" in pa:\n",
    "                    if isinstance(pa['s'], str):\n",
    "                        pa['s'] = df[pa['s']] * 20 + 1\n",
    "                plot = ax.scatter(df['PlotPos'],df[plot['data']], **pa)\n",
    "    \n",
    "    ######## plot the SNP graphs #######\n",
    "    for plot in snp_plots:\n",
    "        if plot['plot_type'] == 'line':\n",
    "            plot = ax.plot(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "        elif plot['plot_type'] == 'scatter':\n",
    "            # highjack plot_args\n",
    "            pa = plot['plot_args']\n",
    "            if \"c\" in pa:\n",
    "                    pa['c'] = df[pa['c']]\n",
    "            if \"s\" in pa:\n",
    "                if isinstance(pa['s'], str):\n",
    "                    pa['s'] = df[pa['s']] * 20 + 1\n",
    "            plot = ax.scatter(df['PlotPos'],df[plot['data']], **pa)\n",
    "    \n",
    "    _ = ax.set_ylim(ylim)\n",
    "    # add the color chroms\n",
    "    _ = make_color_chroms(ax, chrom_df, color_chroms, ylimits=ax.get_ylim(), colormap=colormap)\n",
    "    \n",
    "    \n",
    "    ######## LABELS ###################\n",
    "    # set the axis labels\n",
    "    _ = ax.set_xlabel('genomic coords', fontsize=1.25*label_size);\n",
    "    # quick fix for one y-label\n",
    "    _ = ax.set_ylabel(' / '.join([plot['title'] for plot in snp_plots]), fontsize=1.25*label_size)\n",
    "    \n",
    "    ######## CHROM LABELS #############\n",
    "    add_chrom_labels(ax, chrom_df, ax.get_ylim())\n",
    "    \n",
    "    ####### X-AXIS ####################\n",
    "    # set major ticks and grid for chrom\n",
    "    \n",
    "    ax = set_ticks(ax, df, chrom_df, label_size=label_size)\n",
    "    \n",
    "    # return fig and ax for further plotting and return edited dataframe\n",
    "    return fig, ax, df, chrom_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:53:50.925623Z",
     "start_time": "2020-10-30T20:53:50.445541Z"
    }
   },
   "outputs": [],
   "source": [
    "absvaf = dict(\n",
    "        title='absVAF',\n",
    "        plot_type='scatter',  # ['line', 'scatter']\n",
    "        data='absVAF',\n",
    "        plot_args=dict(\n",
    "            c='dbscan',\n",
    "            s='dbscan',\n",
    "            alpha=.5,\n",
    "            cmap='rainbow'\n",
    "        )\n",
    "    )\n",
    "\n",
    "_ = plot_snp(df, snp_plots=[absvaf, absvafmean], cov_plots=[log2, log2mean], chroms=chroms, region='chr17', **fig_params)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
