{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:11:05.139548Z",
     "start_time": "2020-10-30T20:11:05.134331Z"
    }
   },
   "outputs": [],
   "source": [
    "# HOME\n",
    "import os\n",
    "import scipy\n",
    "home = '/Users/mahtin'\n",
    "# home = '/Users/martinscience'\n",
    "# standard paths\n",
    "testdata = os.path.join(home,\"Dropbox/Icke/Work/somVar/testdata\")\n",
    "static_path = os.path.join(home, \"Dropbox/Icke/Work/static\")\n",
    "cluster_path = os.path.join(home, \"mount\")\n",
    "tooldata = os.path.join(home, \"Dropbox/Icke/Work/somVar/tooldata\")\n",
    "\n",
    "# tool-specific paths\n",
    "shell_path = \"../shell\"\n",
    "# \n",
    "cnvdata = os.path.join(tooldata, \"myCNVdata\")\n",
    "output_path = os.path.join(cnvdata, \"output\")\n",
    "\n",
    "# the path to the input data\n",
    "cnv_path = os.path.join(cnvdata, \"cnv\")\n",
    "cnvPON_path = cnv_path = os.path.join(cnvdata, \"chromCov\")\n",
    "cnv_path = os.path.join(cluster_path, \"scratch/develop/PONcoverage/cnv\")\n",
    "cnvPON_path = os.path.join(cluster_path, \"scratch/develop/PONcoverage/chromCov\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:11:05.700302Z",
     "start_time": "2020-10-30T20:11:05.696652Z"
    }
   },
   "outputs": [],
   "source": [
    "cnvPON_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## visualize coverage data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:11:07.523683Z",
     "start_time": "2020-10-30T20:11:06.756838Z"
    }
   },
   "outputs": [],
   "source": [
    "sample = \"01_A\"\n",
    "cov_df = pd.read_csv(os.path.join(output_path, f'rollingCNV/{sample}.rollcov.csv'), sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### visualize the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:11:09.339886Z",
     "start_time": "2020-10-30T20:11:08.477294Z"
    }
   },
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "# use seaborn plotting defaults\n",
    "import seaborn as sns; sns.set()\n",
    "from matplotlib.collections import PatchCollection\n",
    "from matplotlib.patches import Rectangle\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:11:10.007208Z",
     "start_time": "2020-10-30T20:11:09.981142Z"
    }
   },
   "outputs": [],
   "source": [
    "def sort_df(df):\n",
    "    '''\n",
    "    helper for sorting dfs for chromosomes\n",
    "    '''\n",
    "    df2 = df.copy()\n",
    "    # make Chr column categorical for sorting .. and sort\n",
    "    chrom_list = [f\"chr{i+1}\" for i in range(22)] + ['chrX']\n",
    "    df2['Chr'] = pd.Categorical(df2['Chr'], chrom_list)\n",
    "    return df2.sort_values(['Chr', 'FullExonPos'])\n",
    "\n",
    "\n",
    "def get_chrom_df(df):\n",
    "\n",
    "    # dropna is neccessary because grouping on categorical returns all categories\n",
    "    chrom_df = df.groupby('Chr')['FullExonPos'].agg(['mean', 'min', 'max']).dropna()\n",
    "    cols = list(chrom_df.columns)\n",
    "    chrom_df['sum'] = chrom_df['max'] - chrom_df['min']\n",
    "    chrom_df['cummin'] = chrom_df['sum'].cumsum()\n",
    "    chrom_df['dif'] = (chrom_df['max'] - chrom_df['cummin']).astype(int)\n",
    "    for col in cols:\n",
    "        chrom_df[col] = (chrom_df[col] - chrom_df['dif']).astype(int)\n",
    "    cols.append('dif')\n",
    "    return chrom_df.loc[:, cols]\n",
    "\n",
    "\n",
    "def make_color_chroms(ax, chrom_df, color_chroms, ylimits=(-10,10), colormap='coolwarm_r'):\n",
    "    \n",
    "    # set the cmap from provided argument\n",
    "    cmap = plt.cm.get_cmap(colormap, 23)\n",
    "    \n",
    "    # build the rects\n",
    "    rects = []\n",
    "    # set the height and ymin beyond the ylimits so borders are not seen\n",
    "    ymin = ylimits[0] * 1.1\n",
    "    height = (ylimits[1] - ymin) * 1.1\n",
    "    \n",
    "    for chrom in chrom_df.index:\n",
    "        row = chrom_df.loc[chrom]\n",
    "        rect = Rectangle((row['min'], ymin), width=row['max']-row['min'], height=height)\n",
    "        rects.append(rect)\n",
    "\n",
    "    if color_chroms:\n",
    "        rect_kwargs = dict(\n",
    "            alpha=.6,\n",
    "            ec='none'\n",
    "        )\n",
    "    else:\n",
    "        rect_kwargs = dict(\n",
    "            alpha=1,\n",
    "            fc='none',\n",
    "            ec='darkgray',\n",
    "            lw=1,\n",
    "            ls='-'\n",
    "        )\n",
    "    # set the rectangle collection with colormap\n",
    "    rect_collection = PatchCollection(rects, cmap=cmap, **rect_kwargs)\n",
    "    # set the index for the color map from chrom integers\n",
    "    rect_collection.set_array(chrom_df.index.str.replace('chr', '').str.replace('X', '23').astype(int))\n",
    "    return ax.add_collection(rect_collection)\n",
    "\n",
    "\n",
    "def add_chrom_labels(ax, chrom_df, ylimits=(-10,10)):\n",
    "    \n",
    "    # YOFFSET is the upper-relative y-position \n",
    "    YOFFSET = 0.9\n",
    "    # get the min_chrom_fraction from minimum chrom_size / whole stretch\n",
    "    min_chrom_frac = (chrom_df['max'] - chrom_df['min']).min() / chrom_df['max'].max()\n",
    "    chrom_size = min(50, max(15, 200*min_chrom_frac))\n",
    "    style = dict(size=chrom_size, color='#2f3832')\n",
    "    # set the height and ymin beyond the ylimits so borders are not seen\n",
    "    ypos = ylimits[0] + YOFFSET * (ylimits[1] - ylimits[0])\n",
    "    for chrom, row in chrom_df.iterrows():\n",
    "        if (len(chrom_df.index) > 12):\n",
    "            chrom = chrom.replace('chr', '')\n",
    "        ax.text(row['mean'], ypos, chrom, ha='center', **style)\n",
    "    \n",
    "\n",
    "def make_nice(position):\n",
    "    '''\n",
    "    takes position and returns closest multiple of 1, 2, 5 or 10\n",
    "    '''\n",
    "    # set nice values\n",
    "    nice_positions = np.array([1,2,2.5,5,10])\n",
    "    # get the 10s\n",
    "    power10 = np.power(10, np.floor(np.log10(position)))\n",
    "    # reduce to value between 1 and 10\n",
    "    num = position / power10\n",
    "    # find the closest nice position\n",
    "    base = nice_positions[np.argmin(np.abs(nice_positions / num - 1))]\n",
    "    return base * power10\n",
    "\n",
    "def get_tick_pos(tick_dist, chrom_df):\n",
    "    '''\n",
    "    return from chrom_df the evenly-spread (tick_dist) positions per chrom\n",
    "    '''\n",
    "    return [pos for _, row in chrom_df.iterrows() for pos in range(row['min'] + tick_dist, row['max'], tick_dist)]\n",
    "\n",
    "\n",
    "def str_pos(pos, df, precision=1):\n",
    "    '''\n",
    "    returns string representation of base position\n",
    "    on genomic coords\n",
    "    '''\n",
    "    pos = df.iloc[np.argmin(np.abs(df['PlotPos'] - pos))]['Pos']\n",
    "    # get the closest base power\n",
    "    power10 = int(np.round(np.log10(pos) / 3) * 3)\n",
    "    # get the base fraction\n",
    "    base = pos / np.power(10, power10)\n",
    "    if power10 == 9:\n",
    "        base = base * 1000\n",
    "        power10 = 6\n",
    "    if power10 == 6:\n",
    "        suff = 'Mb'\n",
    "    elif power10 == 3:\n",
    "        suff = 'kb'\n",
    "    base = round(base,precision)\n",
    "    if power10 == 0:\n",
    "        suff = 'b'\n",
    "        base = int(base)\n",
    "    return f\"{base}{suff}\"\n",
    "\n",
    "\n",
    "def set_ticks(ax, df, chrom_df, ticks=20, label_size=12):\n",
    "    '''\n",
    "    for a given tick number, set nicely spread ticks\n",
    "    '''\n",
    "    \n",
    "    ## determine optimale tick distance\n",
    "    # get the chrom_number\n",
    "    chrom_count = len(chrom_df.index)\n",
    "    # get the number of bases\n",
    "    stretch = chrom_df['max'][-1]\n",
    "    # set the number of desired ticks\n",
    "    major_tick_dist = int(stretch / (ticks + 1))\n",
    "    minor_tick_dist = int(stretch / ((ticks * 2) + 1))\n",
    "    \n",
    "    # feed tick distance into chrom_df to get chrom-specific coords\n",
    "    major_pos = get_tick_pos(major_tick_dist, chrom_df)\n",
    "    minor_pos = [pos - minor_tick_dist for pos in major_pos]\n",
    "    \n",
    "    ax.xaxis.set_major_locator(plt.FixedLocator(major_pos))\n",
    "    # only print the genomic coords below a certain base total\n",
    "    if stretch < 2e7:\n",
    "        major_labels = [str_pos(pos, df) for pos in major_pos]\n",
    "        ax.xaxis.set_major_formatter(plt.FixedFormatter(major_labels))\n",
    "    else:\n",
    "        ax.xaxis.set_major_formatter(plt.NullFormatter())\n",
    "    ax.xaxis.set_minor_locator(plt.FixedLocator(minor_pos))\n",
    "    ax.xaxis.grid(which='major', linestyle='-', linewidth=2)\n",
    "    ax.xaxis.grid(which='minor', linestyle='--', linewidth=1)\n",
    "    ax.xaxis.set_tick_params(which='major', length=20, labelsize=label_size)\n",
    "    ax.yaxis.set_tick_params(which='major', length=20, labelsize=label_size)\n",
    "    # set the tick labels\n",
    "    for tick in ax.xaxis.get_majorticklabels():\n",
    "        tick.set_verticalalignment(\"bottom\")\n",
    "    return ax\n",
    "\n",
    "\n",
    "def extract_pos(region):\n",
    "    \n",
    "    def convert(pos):\n",
    "        if pos.endswith('Mb'):\n",
    "            pos = int(pos.replace('Mb', '')) * 1e6\n",
    "        elif pos.endswith('kb'):\n",
    "            pos = int(pos.replace('kb', '')) * 1000\n",
    "        else:\n",
    "            pos = int(pos)\n",
    "        return pos\n",
    "    \n",
    "    split = region.split(':')  \n",
    "    chrom = split[0]\n",
    "    \n",
    "    # if start and are used\n",
    "    if len(split) > 1 and '-' in split[1]:     \n",
    "        se = split[1].split('-')\n",
    "        start = convert(se[0])\n",
    "        end = convert(se[1])\n",
    "    else:\n",
    "        start = 0\n",
    "        end = 1e10\n",
    "    return chrom, start, end\n",
    "    \n",
    "\n",
    "def plot_genomic(df, plots, chroms='all', color_chroms=True, colormap='coolwarm_r', region='', figsize=(20,4), ylim=(-1,1), label_size=12):\n",
    "    \n",
    "    #### DATA MANGELING ##########\n",
    "    # get cols for rearranging\n",
    "    org_cols = list(df.columns)\n",
    "    \n",
    "    # sort the df\n",
    "    df = sort_df(df)\n",
    "    # reduce the df to the selected chromosomes\n",
    "    if region:\n",
    "        chrom, start, end = extract_pos(region)\n",
    "        df = df.query('Chr == @chrom and @start <= Pos <= @end')\n",
    "    elif chroms != 'all':\n",
    "        df = df.query('Chr in @chroms')\n",
    "\n",
    "    # get the chrom_df for collapsing the \n",
    "    chrom_df = get_chrom_df(df)\n",
    "    \n",
    "    df = df.merge(chrom_df.loc[:, 'dif'], on='Chr')\n",
    "    df['PlotPos'] = df['FullExonPos'] - df['dif']\n",
    "    \n",
    "    # rearrange the df as return value\n",
    "    new_cols = org_cols[:4] + ['PlotPos'] + org_cols[4:]\n",
    "    df = df.loc[:, new_cols]\n",
    "    \n",
    "\n",
    "    ######## PLOTTING #######\n",
    "    # plot the figure\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "    # set the x-axis limits\n",
    "    _ = ax.set_xlim(0, df['PlotPos'].max())\n",
    "    \n",
    "    # plot the graphs #######\n",
    "    for plot in plots:\n",
    "        if plot['plot_type'] == 'line':\n",
    "            plot = ax.plot(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "        elif plot['plot_type'] == 'scatter':\n",
    "            plot = ax.scatter(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "    \n",
    "    _ = ax.set_ylim(ylim)\n",
    "    # add the color chroms\n",
    "    _ = make_color_chroms(ax, chrom_df, color_chroms, ylimits=ax.get_ylim(), colormap=colormap)\n",
    "    \n",
    "    \n",
    "    ######## LABELS ###################\n",
    "    # set the axis labels\n",
    "    _ = ax.set_xlabel('genomic coords', fontsize=1.25*label_size);\n",
    "    # quick fix for one y-label\n",
    "    _ = ax.set_ylabel(' / '.join([plot['title'] for plot in plots]), fontsize=1.25*label_size)\n",
    "    \n",
    "    ######## CHROM LABELS #############\n",
    "    add_chrom_labels(ax, chrom_df, ax.get_ylim())\n",
    "    \n",
    "    ####### X-AXIS ####################\n",
    "    # set major ticks and grid for chrom\n",
    "    \n",
    "    ax = set_ticks(ax, df, chrom_df, label_size=label_size)\n",
    "    \n",
    "    # return fig and ax for further plotting and return edited dataframe\n",
    "    return fig, ax, df, chrom_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:11:18.135068Z",
     "start_time": "2020-10-30T20:11:16.815543Z"
    }
   },
   "outputs": [],
   "source": [
    "log2 = dict(\n",
    "        title='log2ratio',\n",
    "        plot_type='scatter',   # ['line', 'scatter']\n",
    "        data='log2ratio',\n",
    "        plot_args=dict(\n",
    "            linewidth=0.3,\n",
    "            color='black',\n",
    "            s=0.2,\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "log2mean = dict(\n",
    "        title='rollinglog2ratio',\n",
    "        plot_type='line',   # ['line', 'scatter']\n",
    "        data='log2ratiomean',\n",
    "        plot_args=dict(\n",
    "            linewidth=1,\n",
    "            color='yellow',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "chroms = ['chr5', 'chr7','chr8', 'chr11', 'chr17']\n",
    "\n",
    "fig_params = dict(\n",
    "    figsize=(24,8),\n",
    "    colormap='coolwarm_r',\n",
    "    color_chroms=True,\n",
    "    ylim=(-1.5,2.5),\n",
    "    label_size=15\n",
    ")\n",
    "\n",
    "fig, ax, df, chrom_df = plot_genomic(cov_df, plots=[log2, log2mean], chroms=chroms, region='', **fig_params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# visualize heteroSNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:11:23.058284Z",
     "start_time": "2020-10-30T20:11:22.639094Z"
    }
   },
   "outputs": [],
   "source": [
    "sample = \"01_A\"\n",
    "snp_df = pd.read_csv(os.path.join(output_path, f'rollingCNV/{sample}.snpcov.csv'), sep='\\t')\n",
    "snp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### for hetSNP visualization:\n",
    "    + covData is normalized to range 1 and pushed up 1.5\n",
    "    + cov_vis is specified by cov_plots\n",
    "    + snp_vis in specified by snp_plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:12:59.975117Z",
     "start_time": "2020-10-30T20:12:59.963092Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_snp(df, snp_plots=[], cov_plots=[], chroms='all', cov_offset=.25, cov_height=0.5, color_chroms=True, colormap='coolwarm_r', region='', label_size=12, figsize=(20,4), ylim=(-1,1)):\n",
    "    \n",
    "    MAXLOG2RATIO = 2.5\n",
    "    #### DATA MANGELING ##########\n",
    "    # get cols for rearranging\n",
    "    org_cols = list(df.columns)\n",
    "    \n",
    "    # sort the df\n",
    "    df = sort_df(df)\n",
    "    # reduce the df to the selected chromosomes\n",
    "    if region:\n",
    "        chrom, start, end = extract_pos(region)\n",
    "        df = df.query('Chr == @chrom and @start <= Pos <= @end')\n",
    "    elif chroms != 'all':\n",
    "        df = df.query('Chr in @chroms')\n",
    "\n",
    "    # get the chrom_df for collapsing the \n",
    "    chrom_df = get_chrom_df(df)\n",
    "    \n",
    "    df = df.merge(chrom_df.loc[:, 'dif'], on='Chr')\n",
    "    df['PlotPos'] = df['FullExonPos'] - df['dif']\n",
    "    \n",
    "    # rearrange the df as return value\n",
    "    new_cols = org_cols[:4] + ['PlotPos'] + org_cols[4:]\n",
    "    df = df.loc[:, new_cols]\n",
    "    \n",
    "    #########################\n",
    "    ######## PLOTTING #######\n",
    "    # plot the figure\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "    # set the x-axis limits\n",
    "    _ = ax.set_xlim(0, df['PlotPos'].max())\n",
    "    \n",
    "    \n",
    "    ######## PLOT COV Data\n",
    "    \n",
    "    if len(cov_plots):\n",
    "        scale_factor = cov_height / (MAXLOG2RATIO + 1)\n",
    "        offset = 1 + scale_factor + cov_offset\n",
    "\n",
    "        ylim = (ylim[0], ylim[1] + cov_offset + cov_height)\n",
    "        \n",
    "        for plot in cov_plots:\n",
    "            # normalize the coverage data:\n",
    "            # 2.5 is the approx max log2ratio (LOH to 8N)\n",
    "\n",
    "            df[plot['data']] = df[plot['data']] * scale_factor + offset\n",
    "            if plot['plot_type'] == 'line':\n",
    "                plot = ax.plot(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "            elif plot['plot_type'] == 'scatter':\n",
    "                plot = ax.scatter(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "    \n",
    "    ######## plot the SNP graphs #######\n",
    "    for plot in snp_plots:\n",
    "        if plot['plot_type'] == 'line':\n",
    "            plot = ax.plot(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "        elif plot['plot_type'] == 'scatter':\n",
    "            plot = ax.scatter(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "    \n",
    "    _ = ax.set_ylim(ylim)\n",
    "    # add the color chroms\n",
    "    _ = make_color_chroms(ax, chrom_df, color_chroms, ylimits=ax.get_ylim(), colormap=colormap)\n",
    "    \n",
    "    \n",
    "    ######## LABELS ###################\n",
    "    # set the axis labels\n",
    "    _ = ax.set_xlabel('genomic coords', fontsize=1.25*label_size);\n",
    "    # quick fix for one y-label\n",
    "    _ = ax.set_ylabel(' / '.join([plot['title'] for plot in snp_plots]), fontsize=1.25*label_size)\n",
    "    \n",
    "    ######## CHROM LABELS #############\n",
    "    add_chrom_labels(ax, chrom_df, ax.get_ylim())\n",
    "    \n",
    "    ####### X-AXIS ####################\n",
    "    # set major ticks and grid for chrom\n",
    "    \n",
    "    ax = set_ticks(ax, df, chrom_df, label_size=label_size)\n",
    "    \n",
    "    # return fig and ax for further plotting and return edited dataframe\n",
    "    return fig, ax, df, chrom_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:48:43.497187Z",
     "start_time": "2020-10-30T20:48:42.742931Z"
    }
   },
   "outputs": [],
   "source": [
    "log2 = dict(\n",
    "        title='log2ratio',\n",
    "        plot_type='scatter',   # ['line', 'scatter']\n",
    "        data='log2ratio',\n",
    "        plot_args=dict(\n",
    "            linewidth=0.3,\n",
    "            color='black',\n",
    "            s=0.2,\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "log2mean = dict(\n",
    "        title='rollinglog2ratio',\n",
    "        plot_type='line',   # ['line', 'scatter']\n",
    "        data='log2ratiomean',\n",
    "        plot_args=dict(\n",
    "            linewidth=1,\n",
    "            color='yellow',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "vaf = dict(\n",
    "        title='VAF',\n",
    "        plot_type='scatter',  # ['line', 'scatter']\n",
    "        data='VAF',\n",
    "        plot_args=dict(\n",
    "            s=.2,\n",
    "            color='black',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "fig_params = dict(\n",
    "    figsize=(24,8),\n",
    "    colormap='coolwarm_r',\n",
    "    color_chroms=True,\n",
    "    ylim=(0,1),\n",
    "    cov_offset=.1,  # how much log2ratio=0 is shifted above SNP-data\n",
    "    cov_height=.5,\n",
    "    label_size=13\n",
    ")\n",
    "\n",
    "fig, ax, df, chrom_df = plot_snp(snp_df, snp_plots=[vaf], cov_plots=[log2, log2mean], chroms=chroms, region='', **fig_params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## visualize clusterings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### load example clustered file from DBscan clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:52:21.378234Z",
     "start_time": "2020-10-30T20:52:21.250918Z"
    }
   },
   "outputs": [],
   "source": [
    "sample = \"01_A\"\n",
    "df = pd.read_csv(os.path.join(output_path, f'cluster/{sample}.dbscan.csv'), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:52:22.176859Z",
     "start_time": "2020-10-30T20:52:22.150268Z"
    }
   },
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:52:23.585512Z",
     "start_time": "2020-10-30T20:52:23.579889Z"
    }
   },
   "outputs": [],
   "source": [
    "log2 = dict(\n",
    "        title='log2ratio',\n",
    "        plot_type='scatter',   # ['line', 'scatter']\n",
    "        data='log2ratio',\n",
    "        plot_args=dict(\n",
    "            linewidth=0.3,\n",
    "            color='black',\n",
    "            s=0.2,\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "log2mean = dict(\n",
    "        title='rollinglog2ratio',\n",
    "        plot_type='line',   # ['line', 'scatter']\n",
    "        data='log2ratiomean',\n",
    "        plot_args=dict(\n",
    "            linewidth=1,\n",
    "            color='yellow',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "absvaf = dict(\n",
    "        title='absVAF',\n",
    "        plot_type='scatter',  # ['line', 'scatter']\n",
    "        data='absVAF',\n",
    "        plot_args=dict(\n",
    "            s=1,\n",
    "            c='dbscan',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "absvafmean = dict(\n",
    "        title='deltaVAFvar',\n",
    "        plot_type='line',   # ['line', 'scatter']\n",
    "        data='deltaVAFvar',\n",
    "        plot_args=dict(\n",
    "            linewidth=1,\n",
    "            color='blue',\n",
    "            alpha=.7\n",
    "        )\n",
    "    )\n",
    "\n",
    "fig_params = dict(\n",
    "    figsize=(24,8),\n",
    "    colormap='coolwarm_r',\n",
    "    color_chroms=False,\n",
    "    ylim=(0,1),\n",
    "    cov_offset=.1,  # how much log2ratio=0 is shifted above SNP-data\n",
    "    cov_height=.5,\n",
    "    label_size=13\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### expand plot_snp functionality to showing clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:52:31.244421Z",
     "start_time": "2020-10-30T20:52:31.230296Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_snp(df, snp_plots=[], cov_plots=[], chroms='all', cov_offset=.25, cov_height=0.5, color_chroms=True, colormap='coolwarm_r', region='', label_size=12, figsize=(20,4), ylim=(-1,1)):\n",
    "    \n",
    "    MAXLOG2RATIO = 2.5\n",
    "    #### DATA MANGELING ##########\n",
    "    # get cols for rearranging\n",
    "    org_cols = list(df.columns)\n",
    "    \n",
    "    # sort the df\n",
    "    df = sort_df(df)\n",
    "    # reduce the df to the selected chromosomes\n",
    "    if region:\n",
    "        chrom, start, end = extract_pos(region)\n",
    "        df = df.query('Chr == @chrom and @start <= Pos <= @end')\n",
    "    elif chroms != 'all':\n",
    "        df = df.query('Chr in @chroms')\n",
    "\n",
    "    # get the chrom_df for collapsing the \n",
    "    chrom_df = get_chrom_df(df)\n",
    "    \n",
    "    df = df.merge(chrom_df.loc[:, 'dif'], on='Chr')\n",
    "    df['PlotPos'] = df['FullExonPos'] - df['dif']\n",
    "    \n",
    "    # rearrange the df as return value\n",
    "    new_cols = org_cols[:4] + ['PlotPos'] + org_cols[4:]\n",
    "    df = df.loc[:, new_cols]\n",
    "    \n",
    "    #########################\n",
    "    ######## PLOTTING #######\n",
    "    # plot the figure\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "    # set the x-axis limits\n",
    "    _ = ax.set_xlim(0, df['PlotPos'].max())\n",
    "    \n",
    "    \n",
    "    ######## PLOT COV Data\n",
    "    \n",
    "    if len(cov_plots):\n",
    "        scale_factor = cov_height / (MAXLOG2RATIO + 1)\n",
    "        offset = 1 + scale_factor + cov_offset\n",
    "\n",
    "        ylim = (ylim[0], ylim[1] + cov_offset + cov_height)\n",
    "        \n",
    "        for plot in cov_plots:\n",
    "            # normalize the coverage data:\n",
    "            # 2.5 is the approx max log2ratio (LOH to 8N)\n",
    "\n",
    "            df[plot['data']] = df[plot['data']] * scale_factor + offset\n",
    "            if plot['plot_type'] == 'line':\n",
    "                plot = ax.plot(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "            elif plot['plot_type'] == 'scatter':\n",
    "                # highjack plot_args\n",
    "                pa = plot['plot_args']\n",
    "                if \"c\" in pa:\n",
    "                    pa['c'] = df[pa['c']]\n",
    "                if \"s\" in pa:\n",
    "                    if isinstance(pa['s'], str):\n",
    "                        pa['s'] = df[pa['s']] * 20 + 1\n",
    "                plot = ax.scatter(df['PlotPos'],df[plot['data']], **pa)\n",
    "    \n",
    "    ######## plot the SNP graphs #######\n",
    "    for plot in snp_plots:\n",
    "        if plot['plot_type'] == 'line':\n",
    "            plot = ax.plot(df['PlotPos'],df[plot['data']], **plot['plot_args'])\n",
    "        elif plot['plot_type'] == 'scatter':\n",
    "            # highjack plot_args\n",
    "            pa = plot['plot_args']\n",
    "            if \"c\" in pa:\n",
    "                    pa['c'] = df[pa['c']]\n",
    "            if \"s\" in pa:\n",
    "                if isinstance(pa['s'], str):\n",
    "                    pa['s'] = df[pa['s']] * 20 + 1\n",
    "            plot = ax.scatter(df['PlotPos'],df[plot['data']], **pa)\n",
    "    \n",
    "    _ = ax.set_ylim(ylim)\n",
    "    # add the color chroms\n",
    "    _ = make_color_chroms(ax, chrom_df, color_chroms, ylimits=ax.get_ylim(), colormap=colormap)\n",
    "    \n",
    "    \n",
    "    ######## LABELS ###################\n",
    "    # set the axis labels\n",
    "    _ = ax.set_xlabel('genomic coords', fontsize=1.25*label_size);\n",
    "    # quick fix for one y-label\n",
    "    _ = ax.set_ylabel(' / '.join([plot['title'] for plot in snp_plots]), fontsize=1.25*label_size)\n",
    "    \n",
    "    ######## CHROM LABELS #############\n",
    "    add_chrom_labels(ax, chrom_df, ax.get_ylim())\n",
    "    \n",
    "    ####### X-AXIS ####################\n",
    "    # set major ticks and grid for chrom\n",
    "    \n",
    "    ax = set_ticks(ax, df, chrom_df, label_size=label_size)\n",
    "    \n",
    "    # return fig and ax for further plotting and return edited dataframe\n",
    "    return fig, ax, df, chrom_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-30T20:53:50.925623Z",
     "start_time": "2020-10-30T20:53:50.445541Z"
    }
   },
   "outputs": [],
   "source": [
    "absvaf = dict(\n",
    "        title='absVAF',\n",
    "        plot_type='scatter',  # ['line', 'scatter']\n",
    "        data='absVAF',\n",
    "        plot_args=dict(\n",
    "            c='dbscan',\n",
    "            s='dbscan',\n",
    "            alpha=.5,\n",
    "            cmap='rainbow'\n",
    "        )\n",
    "    )\n",
    "\n",
    "_ = plot_snp(df, snp_plots=[absvaf, absvafmean], cov_plots=[log2, log2mean], chroms=chroms, region='chr17', **fig_params)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
